# Genome-wide association studies

In this chapter we explain how to perform genome-wide association studies (GWAS), where high-dimensional genomic data is treated. Issues concerning quality control of SNPs and individuals are discussed, as well as association tests and possible sources of confounding such as population stratification. We finish SNP analyses with tools that help to interpret results (e.g. post-omic data analyses). 


GWAS assess the association between the trait of interest and up to millions of SNPs. GWAS have been used to discover thousands of SNPsassociated with several complex diseases [@macarthur2016new]. The basic statistical methods are similar to those previously described, in particular, the massive univariate testing. The main issue with GWAS is data management and computation. Most publicly available data is in [PLINK format](https://www.cog-genomics.org/plink/1.9/input), where genomic data is stored in a binary BED file, and phenotype and annotation data in text BIM and FAM files. PLINK data can be loaded into R with the Bioconductor's package `snpStats`. 

We illustrate the analysis of a GWAS including 100,000 SNPs that have been simulated using real data from a case-control study. Our phenotype of interest is obesity (0: not obese; 1: obese) that has been created using body mass index information of each individual. We start by loading genotype data that are in PLINK format (`obesity.bed`, `obesity.bim`, `obesity.fam` files). 

```{r load_ob}
library(snpStats)
path <- "data"
ob.plink <- read.plink(file.path(path, "obesity"))
```

The imported object is a list containing the genotypes, the family structure and the SNP annotation. 

```{r info_plink}
names(ob.plink)
```

We store genotype, annotation and family data in different variables for downstream analyses

```{r get_geno_annot}
ob.geno <- ob.plink$genotypes
ob.geno

annotation <- ob.plink$map
head(annotation)

family <- ob.plink$fam
head(family)
``` 

Notice that `geno` is an object of class `SnpMatrix`that stores the SNPs in binary (raw) format. While some basic phenotype data is usually available in the `fam` field of the `SnpMatrix` object, a more complete phenotypic characterization of the sample is usually distributed in additional text files. In our example, the complete phenotype data is in a tab-delimited file

```{r read_pheno}
ob.pheno <- read.delim(file.path(path, "obesity.txt"))
head(ob.pheno)
``` 

The file contains phenotypic information for a different set of individuals that overlap with those in the `ob.geno` object. Therefore, before analysis, we need to correctly merge and order the individuals across genomic and phenotype datasets. The row names of `ob.geno` correspond to the individual identifiers (id) variable of `ob.pheno`. Consequently, we also rename the rows of  `ob.pheno` with the `id` variable 

```{r ids}
rownames(ob.pheno) <- ob.pheno$scanID
```

We can check if the row names of the datasets match 

```{r check_ids}
identical(rownames(ob.pheno), rownames(ob.geno))
```

FALSE indicates that either there are different individuals in both objects or that they are in different order. This can be fixed by selecting common individuals.

```{r select_equals}
ids <- intersect(rownames(ob.pheno), rownames(ob.geno))
geno <- ob.geno[ids, ]
ob <- ob.pheno[ids, ]
identical(rownames(ob), rownames(geno))

family <- family[ids, ]
``` 


## Quality control of SNPs

We now perform the quality control (QC)of genomic data at the SNP and individual levels, before association testing [@Anderson]. Different measures can be used to perform QC and remove: 

- SNPs with a high rate of missing; 
- rare SNPS (e.g. having low minor allele frequency (MAF); and
- SNPs that do not pass the HWE test. 

Typically, markers with a call rate less than 95% are removed from association analyses, although some large studies chose higher call-rate thresholds (99%). Markers of low MAF ($<5$%) are also filtered. The significance threshold rejecting a SNP for not being in HWE has varied greatly between studies, from thresholds between 0.001 and $5.7 \times 10^{-7}$ [@clayton2005population]. Including SNPs with extremely low $P$-values for the HWE test will require individual examination of the SNP genotyping process. A parsimonious threshold of 0.001 may be considered, though robustly genotyped SNPs below this threshold may remain in the study [@Anderson], as deviations from HWE\ may indeed arise from biological processes. 

The function `col.summary()` offers different summaries (at SNP level) that can be used in QC

```{r qc_snps}
info.snps <- col.summary(geno)
head(info.snps)
``` 

`snpStats` does not compute $P$-values of the HWE test but computes its $z$-scores. A $P$-value of 0.001 corresponds to a $z$-score of $\pm 3.3$ for a two-tail test. Strictly speaking, the HWE test should be applied to controls only (e.g. obese = 0). However, the default computation is for all samples. 

We thus filter SNPs with a call rate  $>95\%$, MAF of $>5\%$ and $z.HWE<3.3$ in controls 

```{r qc_snp}
controls <- ob$obese ==0 & !is.na(ob$obese)
geno.controls <- geno[controls,]
info.controls <- col.summary(geno.controls)

use <- info.snps$Call.rate > 0.95 &
       info.snps$MAF > 0.05 &
       abs(info.controls$z.HWE < 3.3)    
mask.snps <- use & !is.na(use)

geno.qc.snps <- geno[ , mask.snps]
geno.qc.snps

annotation <- annotation[mask.snps, ]
```

It is common practice to report the number of SNPs that have been removed from the association analyses 

```{r describe_removed_SNPs}
# number of SNPs removed for bad call rate 
sum(info.snps$Call.rate < 0.95)
# number of SNPs removed for low MAF
sum(info.snps$MAF < 0.05, na.rm=TRUE)
#number of SNPs that do not pass HWE test
sum(abs(info.controls$z.HWE > 3.3), na.rm=TRUE)    
# The total number of SNPs do not pass QC
sum(!mask.snps)
```

## Quality control of individuals

QC of individuals, or biological samples, comprises four main steps [@Anderson]: 

- The identification of individuals with discordant reported and genomic sex, 
- the identification of individuals with outlying missing genotype or heterozygosity rate, 
- the identification of duplicated or related individuals, and 
- the identification of individuals of divergent ancestry from the sample. 

We start by removing individuals with sex discrepancies, a large number of missing genotypes and outlying heterozygosity. The function `row.summary()` returns the call rate and the proportion of called SNPs which are heterozygous per individual.

```{r qc_samples}
info.indv <- row.summary(geno.qc.snps)
head(info.indv)
```

Gender is usually inferred from the heterozygosity of chromosome X. Males have an expected heterozygosity of 0 and females of 0.30. Chromosome X heterozygosity can be extracted using `row.summary()` function and then plotted

```{r HetX, fig.cap='Heterozygosity in chromosome X by gender provided in the phenotypic data.'}
geno.X <- geno.qc.snps[,annotation$chromosome=="23" & 
                         !is.na(annotation$chromosome)]
info.X <- row.summary(geno.X)
mycol <- ifelse(ob$gender=="Male", "gray40", "gray80")
plot(info.X$Heterozygosity, col=mycol, 
     pch=16, xlab="Individuals", 
     ylab="Heterozygosity in chromosome X")
legend("topright", c("Males", "Females"), col=mycol,
       pch=16)
```


Figure \@ref(fig:HetX) shows that there are some reported males with non-zero X-heterozygosity and females with zero X-heterozygosity. These samples are located in `sex.discrep` for later removal

```{r sex_disc}
sex.discrep <- (ob$gender=="Male" & info.X$Heterozygosity > 0.2) |  
               (ob$gender=="Female" & info.X$Heterozygosity < 0.2)   
```

Sex filtering based on X-heterozygosity is not sufficient to identify rare aneuploidies, like XXY in males. Alternatively, plots of the mean allelic intensities of SNPs on the X and Y chromosomes can identify mis-annotated sex as well as sex chromosome aneuploidies.


Now, we identify individuals with outlying heterozygosity from the overall genomic heterozygosity rate that is computed by the function `row.summary()`.  Heterozygosity, can also be computed from the statistic $F = 1 - \frac{f(Aa)}{E(f(Aa))}$, where $f(Aa)$ is the observed proportion of heterozygous genotypes (Aa) of a given individual and $E(f(Aa))$ is the expected proportion of heterozygous genotypes. A subject's $E(f(Aa))$ can be computed from the MAF across all the subjects' non-missing SNPs 

```{r compute_FHet}
MAF <- col.summary(geno.qc.snps)$MAF
callmatrix <- !is.na(geno.qc.snps)
hetExp <- callmatrix %*% (2*MAF*(1-MAF))
hetObs <- with(info.indv, Heterozygosity*(ncol(geno.qc.snps))*Call.rate)
info.indv$hetF <- 1-(hetObs/hetExp)

head(info.indv)
```

In Figure \@ref(fig:Het2), we compare $F$ statistic and the Heterozygosity obtained from  `row.summary()`  

```{r Het2, echo=FALSE, fig.cap='Heterozygosity computed using F statistic (left panel) and using row.summary function (right panel). The horizontal dashed line shows a suggestive value to detect individuals with outlier heterozygosity values.'}
par(mfrow=c(1,2))
plot(info.indv$hetF, ylim=c(-0.15,0.15), ylab="F Heterozygosity", 
     pch=16, cex=0.6, col="gray80")
abline(h=0.1, col="red", lty=2)
abline(h=-0.1, col="red", lty=2)
o <- info.indv$hetF>0.1
wordcloud::textplot(c(1:nrow(info.indv))[o], info.indv$hetF[o], rownames(info.indv)[o], new=FALSE, cex=0.8)
plot(info.indv$Heterozygosity, ylab="Heterozygosity", 
     pch=16, cex=0.6, col="gray80")
o <- info.indv$Heterozygosity<0.32
wordcloud::textplot(c(1:nrow(info.indv))[o], info.indv$Heterozygosity[o], rownames(info.indv)[o], new=FALSE, cex=0.8)
```

Individuals whose $F$ statistic is outside the band $\pm 0.1$ are considered sample outliers (left panel Figure \@ref(fig:Het2) and correspond to those having a heterozygosity rate lower than 0.32. 


GWASs are studies that are typically based on population samples. Therefore, close familial relatedness between individuals is not representative of the sample. We, therefore, remove individuals whose relatedness is higher than expected. The package `SNPRelate` is used to perform identity-by-descent (IBD) analysis, computing kinship within the sample. The package requires a data in a GDS format that is obtained with the function `snpgdsBED2GDS`. In addition, IBD analysis requires SNPs that are not in LD (uncorrelated). The function `snpgdsLDpruning` iteratively removes adjacent SNPs that exceed an LD threshold in a sliding window

```{r ibd}
library(SNPRelate)

# Transform PLINK data into GDS format
snpgdsBED2GDS("data/obesity.bed", 
              "data/obesity.fam",
              "data/obesity.bim", 
              out="obGDS")

genofile <- snpgdsOpen("obGDS")

# Prune SNPs for IBD analysis
set.seed(12345)
snps.qc <- colnames(geno.qc.snps)
snp.prune <- snpgdsLDpruning(genofile, ld.threshold = 0.2,
                          snp.id = snps.qc)
snps.ibd <- unlist(snp.prune, use.names=FALSE)
```

Note that this process is performed with SNPs that passed previous QC checks. IBD coefficients are then computed using the method of moments, implemented in the function `snpgdsIBDMoM()`. The result of the analysis is a table indicating kinship among pairs of individuals

```{r ibd2}
ibd <- snpgdsIBDMoM(genofile, kinship=TRUE,
                    snp.id = snps.ibd,
                    num.thread = 1)
ibd.kin <- snpgdsIBDSelection(ibd) 
head(ibd.kin)
``` 

A pair of individuals with higher than expected relatedness are considered with kinship score $> 0.1$

```{r related}
ibd.kin.thres <- subset(ibd.kin, kinship > 0.1)
head(ibd.kin.thres)
```

The ids of the individuals with unusual kinship  are located with `related()` function from the `SNPassoc` package

```{r qc_related}
ids.rel <-  SNPassoc::related(ibd.kin.thres) 
ids.rel
```


Summing up, individuals with more than 5% missing genotypes [@wellcome2007genome,  @silverberg2009ulcerative], with sex discrepancies, $F$ absolute value $>1$ and kinship coefficient $>0.1$ are removed from the genotype and phenotype data

```{r qc_indiv}
use <- info.indv$Call.rate > 0.95 &
       abs(info.indv$hetF) < 0.1 &
       !sex.discrep &
       !rownames(info.indv)%in%ids.rel
mask.indiv <- use & !is.na(use)
geno.qc <- geno.qc.snps[mask.indiv, ]

ob.qc <- ob.pheno[mask.indiv, ]
identical(rownames(ob.qc), rownames(geno.qc))
```

These QC measures are usually reported 

```{r summary_qc}
# number of individuals removed to bad call rate
sum(info.indv$Call.rate < 0.95)
# number of individuals removed for heterozygosity problems
sum(abs(info.indv$hetF) > 0.1)
# number of individuals removed for sex discrepancies
sum(sex.discrep)
# number of individuals removed to be related with others
length(ids.rel)
# The total number of individuals that do not pass QC
sum(!mask.indiv)
```


## Population ancestry

As GWAS are based on general population samples, individual genetic differences between individuals need to be also representative of the population at large. The main source of genetic differences between individuals is ancestry. Therefore, it is important to check that there are not individuals with unexpected genetic differences in the sample. Ancestral differences can be inferred with principal component analysis (PCA) on the genomic data. Individuals with outlying ancestry can be removed from the study while smaller differences in ancestry can be adjusted in the association models, including the first principal components as covariates.   

PCA on genomic data can be computed using the `SNPRelate` package with the `snpgdsPCA()` function. Efficiency can be improved by removing SNPs that are in LD before PCA, see `snps.ibd` object in the previous IBD analysis. In addition the functio  `snpgdsPCA()` allows parallelization with the argument `num.thread` that determines the number of computing cores to be used

```{r pca}
pca <- snpgdsPCA(genofile, sample.id = rownames(geno.qc),  
                           snp.id = snps.ibd, 
                           num.thread=1)
```

A PCA plot for the first two components can be obtained with 

```{r pca_plot, fig.cap="1st and 2nd principal components of obesity GWAS data example."}
with(pca, plot(eigenvect[,1], eigenvect[,2], 
               xlab="1st Principal Component", 
               ylab="2nd Principal Component", 
               main = "Ancestry Plot",
               pch=21, bg="gray90", cex=0.8))
```

Inspection of Figure \@ref(fig:pca_plot) can be used to identify individuals with unusual ancestry and remove them. Individuals with outlying values in the principal components will be considered for QC. In our example, we can see outlying individuals on the right side of the plot with 1st PC $> 0.05$. Smaller differences in ancestry are an important source of bias in association tests, as explained later. Therefore, we keep the first five principal components and add it to the phenotypic information that will be used in the association analyses.


```{r ad_pc}
ob.qc <- data.frame(ob.qc, pca$eigenvect[, 1:5])
```

After performing QC, the GDS file can be closed

```{r close_gds}
closefn.gds(genofile)
```

## Genome-wide association analysis

Genome-wide association analysis involves regressing each SNP separately on our trait of interest. The analyses should be adjusted for clinical, environmental, and/or demographic factors as well as ancestral differences between the subjects. The analysis can be performed with a range of functions in `snpStats` package. We first examine the unadjusted whole genome association of our obesity study

```{r gwas}
res <- single.snp.tests(obese, data=ob.qc, 
                        snp.data=geno.qc)
res[1:5,]
```

This analysis is only available for the additive ($\chi^2$(1.df)) and the codominant models ($\chi^2$(2.df)). It requires the name variable phenotype (obese) in the `data` argument. Genomic data are given in the `snp.data` argument. It is important that the individuals in the rows of both datasets match. SNPs in the mitochondrial genome and gonosomes return NA for the $\chi^2$ estimates. These variants should be analyzed separately. A common interest is to analyze autosomes only, and therefore these SNPs can be removed in the QC process. 

A quantitative trait can also be analyzed setting the argument `family` equal to "Gaussian"

```{r gwas_quant}
res.quant <- snp.rhs.tests(age ~ 1,  data=ob.qc, 
                           snp.data=geno.qc,
                           family="Gaussian")
head(res.quant)
```


Population stratificationinflates the estimates of the $\chi^2$ tests of association between the phenotype and the SNPs, and as a consequence the false positive rate increases. Figure \@ref{fig:stratificationExample} illustrates why population stratification\index{subjectindex}{Population stratification} may lead to false associations. In the hypothetical study in the figure, we compare 20 cases and 20 controls where individuals carrying a susceptibility allele are denoted by a dot. The overall frequency of the susceptibility allele is much larger in cases (0.55 = 11/20) than in controls (0.35 = 7/20), the odds of being a case in allele carriers is $\sim$2.3 times higher than the odds of being a case in non carriers  (OR= 2.27 = (0.55/0.45) / (0.35/0.65)). However, the significant increase in susceptibility between the allele is misleading, as the OR in population A (light color) is 0.89 and in population B (dark color) is 1.08. The susceptibility allele strongly discriminates population A from B, and given the differences of the trait frequency between populations, it is likely that the association of the allele with the trait is through its links with population differences and not with the trait itself. 

\begin{figure}
\centering
\includegraphics[width=10cm,height=8cm]{chapters/Chapter04-Genomic_studies/SNPs/figures/stratification.png}
\caption{Illustrative example of population stratification. Read Section \ref{Section-SNP association} for a detailed description.}
\end{figure}\label{fig:stratificationExample}

In genome-wide analyses, the inflation of the associations due to undetected latent variables is assessed by quantile-quantile (Q-Q) plots where observed $\chi^2$ values are plotted against the expected ones 

```{r qqPlot, fig.cap="QQ-plot corresponding to obesity GWAS data example.", fig.height=3.5, fig.width=3.5}
chi2 <- chi.squared(res, df=1)
qq.chisq(chi2)
```

Figure \@ref{fig:qqPlot} shows, in particular, that the $\chi^2$ estimates are not inflated ($\lambda$ is also close to 1), as all quantile values fall in the confidence bands, meaning that most SNPs are not associated with obesity. In addition, the Figure does not show any top SNP outside the confidence bands. A Q-Q plot with top SNPs outside the confidence bands indicates that those SNPs are truly associated with the disease and, hence, do not follow the null hypothesis. Therefore, the Q-Q plot of our examples reveals no significant SNP associations.  

Q-Q plots are used to inspect population stratification. In particular, when population stratification is present, most SNP Q-Q values will be found outside the confidence bands, suggesting that the overall genetic structure of the sample can discriminate differences between subject traits. The $\lambda$ value is a measure of the degree of inflation. The main source of population stratification that is derived from genomic data is ancestry. Therefore, in the cases of inflated Q-Q plots, it is ancestry differences and not individual SNP differences that explain the differences in the phenotype. Population stratification may be corrected by genomic control, mixed models or EIGENSTRAT method @price2010new. However, the most common approach is to use the inferred ancestry from genomic data as covariates in the association analyses @price2006principal. Genome-wide association analysis typically adjusts for population stratification using the PCs on genomic data to infer ancestral differences in the sample. Covariates are easily incorporated in the model of `snp.rhs.tests()` function.

```{r gwas_adh}
res.adj <- snp.rhs.tests(obese ~ X1 + X2 + X3 + X4 + X5, 
                         data=ob.qc, snp.data=geno.qc)
head(res.adj)
```

This function only computes the additive model, adjusting for the first five genomic PCs. The resulting  $-\log_{10}(P)$-values of association for each SNP are then extracted

```{r p-value}
pval.log10 <- -log10(p.value(res.adj))
```

These transformed $P$-values are used to create a Manhattan plot to visualize which SNPs are significantly associated with obesity. We use our function `manhattanPlot()`, although package `qqman` can also be used.

The function `manhattanPlot()` can be downloaded as follows:

```{r download}
source("https://raw.githubusercontent.com/isglobal-brge/book_omic_association/master/R/manhattanPlot.R")
```


```{r manhattan, fig.cap="Manhattan plot of obesity GWAS data example.", fig.height=6.5, fig.width=9}
library(tidyverse)
library(ggplot2)
library(ggrepel)
# Create the required data frame
pvals <- data.frame(SNP=annotation$snp.name, 
                    CHR=annotation$chromosome,
                    BP=annotation$position,
                    P=p.value(res.adj))
# missing data is not allowed
pvals <- subset(pvals, !is.na(CHR) & !is.na(P)) 

manhattanPlot(pvals, color=c("gray90", "gray40"))
```

Significance at Bonferroni level is set at $10^{-7}=0.05/10^5$, as we are testing 100,000 SNPs. The level corresponds to $-\log_{10}(P)=6.30$. Therefore, we confirm, as expected form the Q-Q plot, that no SNP in our study is significantly associated with obesity, as observed in Figure \@ref{fig:manhattan}. It should be noticed that the standard Bonferroni significant level in GWASs is considered as $5 \times 10^{-8}$ since SNParray data use to contain 500K-1M SNPs [@pe2008estimation].

With our obesity example, we illustrate the common situation of finding no significant associations in small studies (thousands of subjects) with small genomic data (100,000 SNPs). This situation motivates multi-center studies with larger samples sizes, where small effects can be inferred with sufficient power and consistency.  

The `snpStats` package performs association analyses using either codominant or additive models and only provides their p-values. Let us imagine we are interested in using `SNPassoc` to create the tables with the different genetic models for the most significant SNPs. This can be performed as following. We first select the SNPs that pass a p-value threshold. For instance, $10^{-5}$ (in real situations it should be the GWAS level $5 \times 10^{-8}$)

```{r subset_pvals}
topPvals <- subset(pvals, P<10e-5)
topSNPs <- as.character(topPvals$SNP)
```

We then export the data into a text file that can be imported into R as a data.frame that can be analyzed using `SNPassoc`

```{r export_SNPs}
# subset top SNPs
geno.topSNPs <- geno.qc[, topSNPs]
geno.topSNPs
# export top SNPs
write.SnpMatrix(geno.topSNPs, file="topSNPs.txt")
# import top SNPs
ob.top <- read.delim("topSNPs.txt", sep="")
# add phenotypic information(ids are in the same order)
ob.top <- cbind(ob.top, ob.qc)
# prepare data for SNPassoc (SNPs are coded as 0,1,2)
ii <- grep("^rs", names(ob.top))
ob.top.s <- setupSNP(ob.top, colSNPs = ii, 
                     name.genotypes=c(0,1,2))
# run association (all)
WGassociation(obese, ob.top.s)
# run association (one)
association(obese ~ rs10193241, ob.top.s)
```

