<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9 Gene expression-based enrichment analysis | Master in Bioinformatics and Omic Data Analysis</title>
  <meta name="description" content="9 Gene expression-based enrichment analysis | Master in Bioinformatics and Omic Data Analysis" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="9 Gene expression-based enrichment analysis | Master in Bioinformatics and Omic Data Analysis" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9 Gene expression-based enrichment analysis | Master in Bioinformatics and Omic Data Analysis" />
  
  
  

<meta name="author" content="Juan R González" />


<meta name="date" content="2021-01-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="introduction.html"/>
<link rel="next" href="genomic-region-enrichment-analysis.html"/>
<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="bioconductor.html"><a href="bioconductor.html"><i class="fa fa-check"></i><b>2</b> Bioconductor</a>
<ul>
<li class="chapter" data-level="2.1" data-path="bioconductor.html"><a href="bioconductor.html#bioconductors-overview"><i class="fa fa-check"></i><b>2.1</b> Bioconductor’s overview</a></li>
<li class="chapter" data-level="2.2" data-path="bioconductor.html"><a href="bioconductor.html#bioconductor-infrastructures"><i class="fa fa-check"></i><b>2.2</b> Bioconductor infrastructures</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="bioconductor.html"><a href="bioconductor.html#snpmatrix"><i class="fa fa-check"></i><b>2.2.1</b> <code>snpMatrix</code></a></li>
<li class="chapter" data-level="2.2.2" data-path="bioconductor.html"><a href="bioconductor.html#expressionset"><i class="fa fa-check"></i><b>2.2.2</b> <code>ExpressionSet</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="bioconductor.html"><a href="bioconductor.html#summarizedexperiment"><i class="fa fa-check"></i><b>2.2.3</b> <code>SummarizedExperiment</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="bioconductor.html"><a href="bioconductor.html#genomic-ranges-granges"><i class="fa fa-check"></i><b>2.2.4</b> <code>Genomic Ranges (GRanges)</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="bioconductor.html"><a href="bioconductor.html#rangedsummarizedexperiment"><i class="fa fa-check"></i><b>2.2.5</b> <code>RangedSummarizedExperiment</code></a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="bioconductor.html"><a href="bioconductor.html#annotation"><i class="fa fa-check"></i><b>2.3</b> Annotation</a></li>
<li class="chapter" data-level="2.4" data-path="bioconductor.html"><a href="bioconductor.html#annotate-package"><i class="fa fa-check"></i><b>2.4</b> <code>annotate</code> package</a></li>
<li class="chapter" data-level="2.5" data-path="bioconductor.html"><a href="bioconductor.html#biomart"><i class="fa fa-check"></i><b>2.5</b> BioMart</a></li>
<li class="chapter" data-level="2.6" data-path="bioconductor.html"><a href="bioconductor.html#annotate-snps"><i class="fa fa-check"></i><b>2.6</b> Annotate SNPs</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="single-association-analysis.html"><a href="single-association-analysis.html"><i class="fa fa-check"></i><b>3</b> Single association analysis</a>
<ul>
<li class="chapter" data-level="3.1" data-path="single-association-analysis.html"><a href="single-association-analysis.html#descriptive-analysis"><i class="fa fa-check"></i><b>3.1</b> Descriptive analysis</a></li>
<li class="chapter" data-level="3.2" data-path="single-association-analysis.html"><a href="single-association-analysis.html#hardy-weinberg-equilibrium"><i class="fa fa-check"></i><b>3.2</b> Hardy-Weinberg equilibrium</a></li>
<li class="chapter" data-level="3.3" data-path="single-association-analysis.html"><a href="single-association-analysis.html#single-snp-association-analysis"><i class="fa fa-check"></i><b>3.3</b> Single SNP association analysis</a></li>
<li class="chapter" data-level="3.4" data-path="single-association-analysis.html"><a href="single-association-analysis.html#gene-environment-and-gene-gene-interactions"><i class="fa fa-check"></i><b>3.4</b> Gene-environment and gene-gene interactions</a></li>
<li class="chapter" data-level="3.5" data-path="single-association-analysis.html"><a href="single-association-analysis.html#haplotype-analysis"><i class="fa fa-check"></i><b>3.5</b> Haplotype analysis</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="single-association-analysis.html"><a href="single-association-analysis.html#haplotype-estimation"><i class="fa fa-check"></i><b>3.5.1</b> Haplotype estimation</a></li>
<li class="chapter" data-level="3.5.2" data-path="single-association-analysis.html"><a href="single-association-analysis.html#haplotype-association"><i class="fa fa-check"></i><b>3.5.2</b> Haplotype association</a></li>
<li class="chapter" data-level="3.5.3" data-path="single-association-analysis.html"><a href="single-association-analysis.html#sliding-window-approach"><i class="fa fa-check"></i><b>3.5.3</b> Sliding window approach</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="single-association-analysis.html"><a href="single-association-analysis.html#genetic-score"><i class="fa fa-check"></i><b>3.6</b> Genetic score</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html"><i class="fa fa-check"></i><b>4</b> Genome-wide association studies</a>
<ul>
<li class="chapter" data-level="4.1" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#quality-control-of-snps"><i class="fa fa-check"></i><b>4.1</b> Quality control of SNPs</a></li>
<li class="chapter" data-level="4.2" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#quality-control-of-individuals"><i class="fa fa-check"></i><b>4.2</b> Quality control of individuals</a></li>
<li class="chapter" data-level="4.3" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#population-ancestry"><i class="fa fa-check"></i><b>4.3</b> Population ancestry</a></li>
<li class="chapter" data-level="4.4" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#genome-wide-association-analysis"><i class="fa fa-check"></i><b>4.4</b> Genome-wide association analysis</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="microarray-data-analysis.html"><a href="microarray-data-analysis.html"><i class="fa fa-check"></i><b>5</b> Microarray data analysis</a></li>
<li class="chapter" data-level="6" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html"><i class="fa fa-check"></i><b>6</b> RNAseq data analysis</a>
<ul>
<li class="chapter" data-level="6.1" data-path="index.html"><a href="index.html#introduction"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#within-sample-normalization-of-read-counts"><i class="fa fa-check"></i><b>6.2</b> Within sample normalization of read counts</a></li>
<li class="chapter" data-level="6.3" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#exploratory-analysis-of-the-read-counts-table"><i class="fa fa-check"></i><b>6.3</b> Exploratory analysis of the read counts table</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#clustering"><i class="fa fa-check"></i><b>6.3.1</b> Clustering</a></li>
<li class="chapter" data-level="6.3.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#pca"><i class="fa fa-check"></i><b>6.3.2</b> PCA</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.4</b> Differential expression analysis</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#deseq2"><i class="fa fa-check"></i><b>6.4.1</b> DESeq2</a></li>
<li class="chapter" data-level="6.4.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#voom"><i class="fa fa-check"></i><b>6.4.2</b> voom</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#accounting-for-additional-sources-of-variation"><i class="fa fa-check"></i><b>6.5</b> Accounting for additional sources of variation</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#accounting-for-covariates-using-deseq2"><i class="fa fa-check"></i><b>6.5.1</b> Accounting for covariates using <code>DESeq2</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#accounting-for-estimated-covariates-using-ruvseq"><i class="fa fa-check"></i><b>6.5.2</b> Accounting for estimated covariates using RUVSeq</a></li>
<li class="chapter" data-level="6.5.3" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#removing-unwanted-variation-from-the-data"><i class="fa fa-check"></i><b>6.5.3</b> Removing unwanted variation from the data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html"><i class="fa fa-check"></i><b>7</b> Enrichment Analysis</a></li>
<li class="chapter" data-level="8" data-path="index.html"><a href="index.html#introduction"><i class="fa fa-check"></i><b>8</b> Introduction</a></li>
<li class="chapter" data-level="9" data-path="gene-expression-based-enrichment-analysis.html"><a href="gene-expression-based-enrichment-analysis.html"><i class="fa fa-check"></i><b>9</b> Gene expression-based enrichment analysis</a>
<ul>
<li class="chapter" data-level="9.1" data-path="gene-expression-based-enrichment-analysis.html"><a href="gene-expression-based-enrichment-analysis.html#a-primer-on-terminology-existing-methods-statistical-theory"><i class="fa fa-check"></i><b>9.1</b> A primer on terminology, existing methods &amp; statistical theory</a></li>
<li class="chapter" data-level="9.2" data-path="gene-expression-based-enrichment-analysis.html"><a href="gene-expression-based-enrichment-analysis.html#basic-analysis-using-gostats"><i class="fa fa-check"></i><b>9.2</b> Basic analysis using <code>GOstats</code></a></li>
<li class="chapter" data-level="9.3" data-path="gene-expression-based-enrichment-analysis.html"><a href="gene-expression-based-enrichment-analysis.html#analysis-using-clusterprofiler"><i class="fa fa-check"></i><b>9.3</b> Analysis using <code>clusterProfiler</code></a></li>
<li class="chapter" data-level="9.4" data-path="gene-expression-based-enrichment-analysis.html"><a href="gene-expression-based-enrichment-analysis.html#visualization"><i class="fa fa-check"></i><b>9.4</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="genomic-region-enrichment-analysis.html"><a href="genomic-region-enrichment-analysis.html"><i class="fa fa-check"></i><b>10</b> Genomic region enrichment analysis</a></li>
<li class="chapter" data-level="11" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>11</b> References</a>
<ul>
<li class="chapter" data-level="11.1" data-path="references.html"><a href="references.html#BioCEx"><i class="fa fa-check"></i><b>11.1</b> Bioconductor</a></li>
<li class="chapter" data-level="11.2" data-path="references.html"><a href="references.html#genetic-association-studies"><i class="fa fa-check"></i><b>11.2</b> Genetic association studies</a></li>
<li class="chapter" data-level="11.3" data-path="references.html"><a href="references.html#gwas"><i class="fa fa-check"></i><b>11.3</b> GWAS</a></li>
<li class="chapter" data-level="11.4" data-path="references.html"><a href="references.html#Microarray-exerc"><i class="fa fa-check"></i><b>11.4</b> Microarray data analysis</a></li>
<li class="chapter" data-level="11.5" data-path="references.html"><a href="references.html#RNAseq-exerc"><i class="fa fa-check"></i><b>11.5</b> RNAseq data analysis</a></li>
<li class="chapter" data-level="11.6" data-path="references.html"><a href="references.html#enrichment-exerc"><i class="fa fa-check"></i><b>11.6</b> Enrichment Analysis</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Master in Bioinformatics and Omic Data Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="gene-expression-based-enrichment-analysis" class="section level1" number="9">
<h1><span class="header-section-number">9</span> Gene expression-based enrichment analysis</h1>
<p>This methodology can be extended to the analysis of multiple gene sets available in different databases such as GO or KEGG among others. Next sections illustrate how to perform such analyses using different BioC packages.</p>
<div id="a-primer-on-terminology-existing-methods-statistical-theory" class="section level2" number="9.1">
<h2><span class="header-section-number">9.1</span> A primer on terminology, existing methods &amp; statistical theory</h2>
<p><strong>Gene sets, pathways &amp; regulatory networks</strong></p>
<p>Gene sets are simple lists of usually functionally related genes without further
specification of relationships between genes.</p>
<p>Pathways can be interpreted as specific gene sets, typically representing a
group of genes that work together in a biological process. Pathways are commonly divided in metabolic and signaling pathways.
Metabolic pathways such as glycolysis represent biochemical substrate conversions
by specific enzymes. Signaling pathways such as the MAPK signaling pathway describe
signal transduction cascades from receptor proteins to transcription factors,
resulting in activation or inhibition of specific target genes.</p>
<p>Gene regulatory networks describe the interplay and effects of regulatory
factors (such as transcription factors and microRNAs) on the expression of their
target genes.</p>
<p><strong>Resources</strong></p>
<p><a href="http://www.geneontology.org">GO</a>, <a href="http://www.genome.jp/kegg">KEGG</a> and <a href="http://software.broadinstitute.org/gsea/msigdb/collections.jsp">MSigDB</a>
annotations are most frequently used for the enrichment analysis of
functional gene sets. Despite an increasing number of gene set and pathway
databases, they are typically the first choice due to their long-standing
curation and availability for a wide range of species.</p>
<p><em>GO</em>: The Gene Ontology (GO) consists of three major sub-ontologies that classify
gene products according to molecular function (MF), biological process (BP) and
cellular component (CC). Each ontology consists of GO terms that define MFs, BPs
or CCs to which specific genes are annotated. The terms are organized in a
directed acyclic graph, where edges between the terms represent
relationships of different types. They relate the terms according to a parent-child
scheme, i.e. parent terms denote more general entities, whereas child terms represent
more specific entities.</p>
<p><em>KEGG</em>: The Kyoto Encyclopedia of Genes and Genomes (KEGG) is a collection of
manually drawn pathway maps representing molecular interaction and reaction networks.
These pathways cover a wide range of biochemical processes that can be divided in
7 broad categories: metabolism, genetic and environmental information processing,
cellular processes, organismal systems, human diseases, and drug development.
Metabolism and drug development pathways differ from pathways of the other 5
categories by illustrating reactions between chemical compounds.
Pathways of the other 5 categories illustrate molecular interactions between
genes and gene products.</p>
<p><em>MSigDB</em>: Molecular Signatures Database (MSigDB) are divided into 8 major collections, and several sub-collections. These include:</p>
<ul>
<li><p><em>H: hallmark gene sets</em>: Hallmark gene sets summarize and represent specific well-defined biological states or processes and display coherent expression.</p></li>
<li><p><em>C1: positional gene sets</em>: Gene sets corresponding to each human chromosome and each cytogenetic band that has at least one gene.</p></li>
<li><p><em>C2: curated gene sets</em>: Gene sets curated from various sources such as online pathway databases, the biomedical literature, and knowledge of domain experts. It contains 5 sub-collections.</p></li>
<li><p><em>C3 motif gene sets</em>: Gene sets representing potential targets of regulation by transcription factors or microRNAs. It has 2 sub-collections.</p></li>
<li><p><em>C4 computational gene sets</em>: Computational gene sets defined by mining large collections of cancer-oriented microarray data. It has 4 sub-collections.</p></li>
<li><p><em>C5 GO gene sets</em>: Gene sets that contain genes annotated by the same GO term. The C5 collection is divided into three sub-collections based on GO ontologies: BP, CC, and MF.</p></li>
<li><p><em>C6 oncogenic signatures</em>: Gene sets that represent signatures of cellular pathways which are often dis-regulated in cancer.</p></li>
<li><p><em>C7 inmunologic signatures</em>: Gene sets that represent cell states and perturbations within the immune system.</p></li>
</ul>
<p><strong>Statistical approaches</strong></p>
<p>The two predominantly used enrichment methods are:</p>
<ul>
<li>Overrepresentation analysis (ORA), testing whether a gene set contains
disproportional many genes of significant expression change, based on the
procedure outlined in the first section.</li>
<li>Gene set enrichment analysis (GSEA), testing whether genes of a gene set
accumulate at the top or bottom of the full gene vector ordered by direction
and magnitude of expression change
<a href="https://doi.org/10.1073/pnas.0506580102">Subramanian et al., 2005</a>. However, the term <em>gene set enrichment analysis</em> nowadays subsumes a general
strategy implemented by a wide range of methods
<a href="https://doi.org/10.1093/nar/gkn923">Huang et al., 2009</a>.
Those methods have in common the same goal, although approach and statistical
model can vary substantially
<a href="https://doi.org/10.1093/bioinformatics/btm051">Goeman and Buehlmann, 2007</a>,
<a href="https://doi.org/10.1371/journal.pcbi.1002375">Khatri et al., 2012</a>.</li>
</ul>
<p>To better distinguish from the specific method, some authors use the term
<em>gene set analysis</em> to denote the general strategy.
However, there is also a specific method from
<a href="https://doi.org/10.1214/07-AOAS101">Efron and Tibshirani, 2007</a> of this name.</p>
<p>Goeman and Buehlmann further raise several critical issues concerning the 2x2 ORA:</p>
<ul>
<li>rather arbitrary classification of genes in DE / not DE</li>
<li>based on gene sampling, although sampling of subjects is appropriate</li>
<li>unrealistic independence assumption between genes, resulting in highly
anti-conservative <em>p</em>-values</li>
</ul>
<p>With regard to these statistical concerns, GSEA is considered superior:</p>
<ul>
<li>takes all measured genes into account</li>
<li>subject sampling via permutation of class labels<br />
</li>
<li>the incorporated permutation procedure implicitly accounts for correlations
between genes</li>
</ul>
<p>However, the simplicity and general applicability of ORA is unmet by subsequent
methods improving on these issues. For instance, GSEA requires the expression data
as input, which is not available for gene lists derived from other experiment types.
On the other hand, the involved sample permutation procedure has been proven
inaccurate and time-consuming
<a href="https://doi.org/10.1214/07-AOAS101">Efron and Tibshirani, 2007</a>,
<a href="https://doi.org/10.2202/1544-6115.1585">Phipson and Smyth, 2010</a>,
<a href="https://doi.org/10.1186/s12859-015-0571-7">Larson and Owen, 2015</a>.</p>
<p><a href="https://doi.org/10.1371/journal.pcbi.1002375">Khatri et al., 2012</a> have taken a
slightly different approach by classifying methods along the timeline of
development into three generations:</p>
<ol style="list-style-type: decimal">
<li>Generation: ORA methods based on the 2x2 contingency table test,</li>
<li>Generation: functional class scoring (FCS) methods such as GSEA, which compute
gene set (= functional class) scores by summarizing per-gene DE statistics,</li>
<li>Generation: topology-based methods, explicitly taking into account interactions
between genes as defined in signaling pathways and gene regulatory networks
(<a href="https://doi.org/10.1093/bioinformatics/btr228">Geistlinger et al., 2011</a> for an example).</li>
</ol>
<p>Although topology-based (also: network-based) methods appear to be most realistic,
their straightforward application can be impaired by features that are not-detectable
on the transcriptional level (such as protein-protein interactions) and insufficient network knowledge
<a href="https://doi.org/10.1093/nar/gkt631">Geistlinger et al., 2013</a>,
<a href="https://doi.org/10.1186/s12859-015-0751-5">Bayerlova et al., 2015</a>.</p>
<p>Given the individual benefits and limitations of existing methods,
cautious interpretation of results is required to derive valid conclusions.
Whereas no single method is best suited for all application scenarios, applying multiple methods can be beneficial.
This has been shown to filter out spurious hits of individual methods, thereby reducing the outcome to gene sets accumulating evidence from different methods
<a href="https://doi.org/10.1186/s12859-016-0884-1">Geistlinger et al., 2016</a>,
<a href="https://doi.org/10.1093/bioinformatics/btw623">Alhamdoosh et al., 2017</a>.</p>
</div>
<div id="basic-analysis-using-gostats" class="section level2" number="9.2">
<h2><span class="header-section-number">9.2</span> Basic analysis using <code>GOstats</code></h2>
<p>Let us illustrate how to perform ORA (enrichment analysis) using gene expression data. First, we need to perform DE analysis. Our interest is to compare gene expression between cell lines that have been treated with dexamethasone or not.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="gene-expression-based-enrichment-analysis.html#cb5-1"></a><span class="kw">library</span>(airway)</span></code></pre></div>
<pre><code>## Loading required package: SummarizedExperiment</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## Loading required package: parallel</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:parallel&#39;:
## 
##     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
##     clusterExport, clusterMap, parApply, parCapply, parLapply,
##     parLapplyLB, parRapply, parSapply, parSapplyLB</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, append, as.data.frame, basename, cbind, colnames,
##     dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
##     grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
##     order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
##     rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
##     union, unique, unsplit, which, which.max, which.min</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     expand.grid</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## 
## Attaching package: &#39;IRanges&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:grDevices&#39;:
## 
##     windows</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## Loading required package: DelayedArray</code></pre>
<pre><code>## Loading required package: matrixStats</code></pre>
<pre><code>## Warning: package &#39;matrixStats&#39; was built under R version 4.0.3</code></pre>
<pre><code>## 
## Attaching package: &#39;matrixStats&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:Biobase&#39;:
## 
##     anyMissing, rowMedians</code></pre>
<pre><code>## 
## Attaching package: &#39;DelayedArray&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:matrixStats&#39;:
## 
##     colMaxs, colMins, colRanges, rowMaxs, rowMins, rowRanges</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     aperm, apply, rowsum</code></pre>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="gene-expression-based-enrichment-analysis.html#cb32-1"></a><span class="kw">data</span>(airway)</span>
<span id="cb32-2"><a href="gene-expression-based-enrichment-analysis.html#cb32-2"></a>airway</span></code></pre></div>
<pre><code>## class: RangedSummarizedExperiment 
## dim: 64102 8 
## metadata(1): &#39;&#39;
## assays(1): counts
## rownames(64102): ENSG00000000003 ENSG00000000005 ... LRG_98 LRG_99
## rowData names(0):
## colnames(8): SRR1039508 SRR1039509 ... SRR1039520 SRR1039521
## colData names(9): SampleName cell ... Sample BioSample</code></pre>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="gene-expression-based-enrichment-analysis.html#cb34-1"></a><span class="kw">table</span>(airway<span class="op">$</span>dex)</span></code></pre></div>
<pre><code>## 
##   trt untrt 
##     4     4</code></pre>
<p>Then, let us analyzed the RNA-seq count data using <code>DESeq2</code> that uses generalized linear models (negative binomial) (<code>voom</code>, <code>edgeR</code> or any other method could be used instead). <strong>NOTES:</strong> 1) we avoid normalization step which is normally required (<code>DESeq2</code> incorporates this step in its pipeline); 2) we also avoid to investigate whether unwanted variability should be removed; 3) Our aim is to get a list of DE genes that may have ‘some sense’</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="gene-expression-based-enrichment-analysis.html#cb36-1"></a><span class="kw">library</span>(DESeq2)</span>
<span id="cb36-2"><a href="gene-expression-based-enrichment-analysis.html#cb36-2"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeqDataSetFromMatrix</span>(<span class="dt">countData =</span> <span class="kw">assay</span>(airway, <span class="st">&quot;counts&quot;</span>), </span>
<span id="cb36-3"><a href="gene-expression-based-enrichment-analysis.html#cb36-3"></a>                              <span class="dt">colData =</span> <span class="kw">colData</span>(airway), </span>
<span id="cb36-4"><a href="gene-expression-based-enrichment-analysis.html#cb36-4"></a>                              <span class="dt">design =</span> <span class="kw">as.formula</span>( <span class="op">~</span><span class="st"> </span>dex))</span>
<span id="cb36-5"><a href="gene-expression-based-enrichment-analysis.html#cb36-5"></a>dds &lt;-<span class="st"> </span>dds[ <span class="kw">rowSums</span>(<span class="kw">counts</span>(dds)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">5</span>, ]</span>
<span id="cb36-6"><a href="gene-expression-based-enrichment-analysis.html#cb36-6"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeq</span>(dds)</span></code></pre></div>
<pre><code>## estimating size factors</code></pre>
<pre><code>## estimating dispersions</code></pre>
<pre><code>## gene-wise dispersion estimates</code></pre>
<pre><code>## mean-dispersion relationship</code></pre>
<pre><code>## final dispersion estimates</code></pre>
<pre><code>## fitting model and testing</code></pre>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="gene-expression-based-enrichment-analysis.html#cb43-1"></a>res &lt;-<span class="st"> </span><span class="kw">results</span>(dds, <span class="dt">contrast =</span> <span class="kw">c</span>(<span class="st">&quot;dex&quot;</span>, <span class="st">&#39;trt&#39;</span>, <span class="st">&#39;untrt&#39;</span>))</span>
<span id="cb43-2"><a href="gene-expression-based-enrichment-analysis.html#cb43-2"></a>res &lt;-<span class="st"> </span>res[<span class="kw">order</span>(res<span class="op">$</span>pvalue), ]</span>
<span id="cb43-3"><a href="gene-expression-based-enrichment-analysis.html#cb43-3"></a>res</span></code></pre></div>
<pre><code>## log2 fold change (MLE): dex trt vs untrt 
## Wald test p-value: dex trt vs untrt 
## DataFrame with 24345 rows and 6 columns
##                  baseMean log2FoldChange     lfcSE       stat       pvalue
##                 &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt;    &lt;numeric&gt;
## ENSG00000152583   997.440        4.60253 0.2117509    21.7356 9.46071e-105
## ENSG00000148175 11193.719        1.45147 0.0852982    17.0164  6.20890e-65
## ENSG00000179094   776.597        3.18386 0.2015137    15.7997  3.12585e-56
## ENSG00000134686  2737.982        1.38715 0.0920523    15.0691  2.58489e-51
## ENSG00000125148  3656.253        2.20344 0.1475941    14.9290  2.13337e-50
## ...                   ...            ...       ...        ...          ...
## ENSG00000258486  330.4721     -0.1421492  0.943248 -0.1507019           NA
## ENSG00000260910   26.1850      1.6795123  0.821286  2.0449778           NA
## ENSG00000261239   10.2232     -0.2123196  2.486879 -0.0853759           NA
## ENSG00000262902  439.5753      0.2233259  1.970163  0.1133540           NA
## ENSG00000265735  143.3031     -0.0906931  1.001134 -0.0905904           NA
##                         padj
##                    &lt;numeric&gt;
## ENSG00000152583 1.71787e-100
## ENSG00000148175  5.63706e-61
## ENSG00000179094  1.89197e-52
## ENSG00000134686  1.17341e-47
## ENSG00000125148  7.74754e-47
## ...                      ...
## ENSG00000258486           NA
## ENSG00000260910           NA
## ENSG00000261239           NA
## ENSG00000262902           NA
## ENSG00000265735           NA</code></pre>
<p>We call DE genes with a minimum 1.2-fold change of expression at
a maximum FDR of 1%:</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="gene-expression-based-enrichment-analysis.html#cb45-1"></a>mask &lt;-<span class="st"> </span>res<span class="op">$</span>padj <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span> <span class="op">&amp;</span><span class="st"> </span><span class="op">!</span><span class="kw">is.na</span>(res<span class="op">$</span>padj) <span class="op">&amp;</span></span>
<span id="cb45-2"><a href="gene-expression-based-enrichment-analysis.html#cb45-2"></a><span class="st">        </span><span class="kw">abs</span>(res<span class="op">$</span>log2FoldChange) <span class="op">&gt;</span><span class="st"> </span><span class="kw">log2</span>(<span class="dv">2</span>)</span>
<span id="cb45-3"><a href="gene-expression-based-enrichment-analysis.html#cb45-3"></a>deGenes &lt;-<span class="st"> </span><span class="kw">rownames</span>(res[mask, ])</span>
<span id="cb45-4"><a href="gene-expression-based-enrichment-analysis.html#cb45-4"></a><span class="kw">head</span>(deGenes)</span></code></pre></div>
<pre><code>## [1] &quot;ENSG00000152583&quot; &quot;ENSG00000148175&quot; &quot;ENSG00000179094&quot; &quot;ENSG00000134686&quot;
## [5] &quot;ENSG00000125148&quot; &quot;ENSG00000120129&quot;</code></pre>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="gene-expression-based-enrichment-analysis.html#cb47-1"></a><span class="kw">length</span>(deGenes)</span></code></pre></div>
<pre><code>## [1] 676</code></pre>
<p>The gene universe is obtained by (number of calculated tests):</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="gene-expression-based-enrichment-analysis.html#cb49-1"></a>geneUniverse &lt;-<span class="st"> </span><span class="kw">rownames</span>(res[<span class="op">!</span><span class="kw">is.na</span>(res<span class="op">$</span>pvalue), ])</span>
<span id="cb49-2"><a href="gene-expression-based-enrichment-analysis.html#cb49-2"></a><span class="kw">length</span>(geneUniverse)</span></code></pre></div>
<pre><code>## [1] 24294</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="gene-expression-based-enrichment-analysis.html#cb51-1"></a><span class="kw">nrow</span>(res)</span></code></pre></div>
<pre><code>## [1] 24345</code></pre>
<p>In order to asses functional enrichment, both DE gene list and gene universe must be annotated in Entrez IDs:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="gene-expression-based-enrichment-analysis.html#cb53-1"></a><span class="kw">library</span>(org.Hs.eg.db)</span></code></pre></div>
<pre><code>## Loading required package: AnnotationDbi</code></pre>
<pre><code>## </code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="gene-expression-based-enrichment-analysis.html#cb56-1"></a>deGenes &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">mget</span>(deGenes, <span class="dt">envir=</span>org.Hs.egENSEMBL2EG,</span>
<span id="cb56-2"><a href="gene-expression-based-enrichment-analysis.html#cb56-2"></a>                       <span class="dt">ifnotfound =</span> <span class="ot">NA</span>))</span>
<span id="cb56-3"><a href="gene-expression-based-enrichment-analysis.html#cb56-3"></a></span>
<span id="cb56-4"><a href="gene-expression-based-enrichment-analysis.html#cb56-4"></a>geneUniverse &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">mget</span>(geneUniverse, <span class="dt">envir=</span>org.Hs.egENSEMBL2EG,</span>
<span id="cb56-5"><a href="gene-expression-based-enrichment-analysis.html#cb56-5"></a>                       <span class="dt">ifnotfound =</span> <span class="ot">NA</span>))</span></code></pre></div>
<p>Now, the enrichment analysis for GO is performed by</p>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="gene-expression-based-enrichment-analysis.html#cb57-1"></a><span class="kw">library</span>(GOstats)</span></code></pre></div>
<pre><code>## Loading required package: Category</code></pre>
<pre><code>## Loading required package: Matrix</code></pre>
<pre><code>## 
## Attaching package: &#39;Matrix&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     expand</code></pre>
<pre><code>## Loading required package: graph</code></pre>
<pre><code>## </code></pre>
<pre><code>## 
## Attaching package: &#39;GOstats&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:AnnotationDbi&#39;:
## 
##     makeGOGraph</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="gene-expression-based-enrichment-analysis.html#cb66-1"></a>params &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;GOHyperGParams&quot;</span>, <span class="dt">geneIds=</span>deGenes,</span>
<span id="cb66-2"><a href="gene-expression-based-enrichment-analysis.html#cb66-2"></a>              <span class="dt">universeGeneIds=</span>geneUniverse,</span>
<span id="cb66-3"><a href="gene-expression-based-enrichment-analysis.html#cb66-3"></a>              <span class="dt">annotation=</span><span class="st">&quot;org.Hs.eg.db&quot;</span>, <span class="dt">ontology=</span><span class="st">&quot;BP&quot;</span>,</span>
<span id="cb66-4"><a href="gene-expression-based-enrichment-analysis.html#cb66-4"></a>              <span class="dt">pvalueCutoff=</span><span class="fl">0.05</span>, <span class="dt">conditional=</span><span class="ot">FALSE</span>,</span>
<span id="cb66-5"><a href="gene-expression-based-enrichment-analysis.html#cb66-5"></a>              <span class="dt">testDirection=</span><span class="st">&quot;over&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in makeValidParams(.Object): removing duplicate IDs in geneIds</code></pre>
<pre><code>## Warning in makeValidParams(.Object): removing duplicate IDs in universeGeneIds</code></pre>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="gene-expression-based-enrichment-analysis.html#cb69-1"></a>hgOver &lt;-<span class="st"> </span><span class="kw">hyperGTest</span>(params)</span>
<span id="cb69-2"><a href="gene-expression-based-enrichment-analysis.html#cb69-2"></a>hgOver</span></code></pre></div>
<pre><code>## Gene to GO BP  test for over-representation 
## 6327 GO BP ids tested (1513 have p &lt; 0.05)
## Selected gene set size: 547 
##     Gene universe size: 13655 
##     Annotation package: org.Hs.eg</code></pre>
<p>The table can be generated by</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="gene-expression-based-enrichment-analysis.html#cb71-1"></a><span class="kw">head</span>(<span class="kw">summary</span>(hgOver))</span></code></pre></div>
<pre><code>##       GOBPID       Pvalue OddsRatio  ExpCount Count Size
## 1 GO:0050896 4.004729e-13  1.901455 271.99780   354 6790
## 2 GO:0009653 1.305943e-12  2.080880  87.36778   151 2181
## 3 GO:0051239 1.464489e-12  2.020228 100.50699   167 2509
## 4 GO:0051094 1.936645e-12  2.459144  44.62527    94 1114
## 5 GO:0048646 1.671464e-11  2.542808  35.85244    79  895
## 6 GO:0050793 2.019617e-11  2.017808  84.76397   144 2116
##                                                       Term
## 1                                     response to stimulus
## 2                       anatomical structure morphogenesis
## 3           regulation of multicellular organismal process
## 4             positive regulation of developmental process
## 5 anatomical structure formation involved in morphogenesis
## 6                      regulation of developmental process</code></pre>
<p>We can alos explore the results by generating an html report:</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="gene-expression-based-enrichment-analysis.html#cb73-1"></a><span class="kw">htmlReport</span>(hgOver, <span class="dt">file=</span><span class="st">&quot;goBPuncond.html&quot;</span>)</span></code></pre></div>
<p>The KEGG enrichment is performed by:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="gene-expression-based-enrichment-analysis.html#cb74-1"></a><span class="kw">library</span>(KEGG.db)</span></code></pre></div>
<pre><code>## 
## KEGG.db contains mappings based on older data because the original
##   resource was removed from the the public domain before the most
##   recent update was produced. This package should now be considered
##   deprecated and future versions of Bioconductor may not have it
##   available.  Users who want more current data are encouraged to look
##   at the KEGGREST or reactome.db packages</code></pre>
<pre><code>## Warning: Package &#39;KEGG.db&#39; is deprecated and will be removed from Bioconductor
##   version 3.12</code></pre>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="gene-expression-based-enrichment-analysis.html#cb77-1"></a>params.kegg &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;KEGGHyperGParams&quot;</span>, <span class="dt">geneIds=</span>deGenes,</span>
<span id="cb77-2"><a href="gene-expression-based-enrichment-analysis.html#cb77-2"></a>                   <span class="dt">universeGeneIds=</span>geneUniverse,</span>
<span id="cb77-3"><a href="gene-expression-based-enrichment-analysis.html#cb77-3"></a>                   <span class="dt">annotation=</span><span class="st">&quot;org.Hs.eg.db&quot;</span>, </span>
<span id="cb77-4"><a href="gene-expression-based-enrichment-analysis.html#cb77-4"></a>                   <span class="dt">pvalueCutoff=</span><span class="fl">0.05</span>, </span>
<span id="cb77-5"><a href="gene-expression-based-enrichment-analysis.html#cb77-5"></a>                   <span class="dt">testDirection=</span><span class="st">&quot;over&quot;</span>)</span></code></pre></div>
<pre><code>## Warning in makeValidParams(.Object): removing duplicate IDs in geneIds</code></pre>
<pre><code>## Warning in makeValidParams(.Object): removing duplicate IDs in universeGeneIds</code></pre>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="gene-expression-based-enrichment-analysis.html#cb80-1"></a>hgOver.kegg &lt;-<span class="st"> </span><span class="kw">hyperGTest</span>(params.kegg)</span>
<span id="cb80-2"><a href="gene-expression-based-enrichment-analysis.html#cb80-2"></a><span class="kw">head</span>(<span class="kw">summary</span>(hgOver.kegg))</span></code></pre></div>
<pre><code>##   KEGGID       Pvalue OddsRatio  ExpCount Count Size
## 1  04060 0.0009526255  2.519903 8.0269137    18  161
## 2  04977 0.0014798039  7.473025 0.8974189     5   18
## 3  00750 0.0021867795 19.300448 0.2991396     3    6
## 4  05414 0.0030336333  3.118754 3.6395323    10   73
## 5  04512 0.0044995301  2.929795 3.8389587    10   77
## 6  04350 0.0049414310  2.886029 3.8888154    10   78
##                                     Term
## 1 Cytokine-cytokine receptor interaction
## 2       Vitamin digestion and absorption
## 3                  Vitamin B6 metabolism
## 4                 Dilated cardiomyopathy
## 5               ECM-receptor interaction
## 6             TGF-beta signaling pathway</code></pre>
</div>
<div id="analysis-using-clusterprofiler" class="section level2" number="9.3">
<h2><span class="header-section-number">9.3</span> Analysis using <code>clusterProfiler</code></h2>
<p><a href="">clusterProfiler</a> performs statistical analysis and visualization of functional profiles for genes and gene clusters. It includes GO, KEEG and a general function to perform ORA for any database such as <a href="http://www.gsea-msigdb.org/gsea/msigdb/index.jsp">MSigDB</a> or <a href="http://www.disgenet.org/">DisGeNET</a> among others.</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="gene-expression-based-enrichment-analysis.html#cb82-1"></a><span class="kw">library</span>(clusterProfiler)</span></code></pre></div>
<pre><code>## clusterProfiler v3.16.1  For help: https://guangchuangyu.github.io/software/clusterProfiler
## 
## If you use clusterProfiler in published research, please cite:
## Guangchuang Yu, Li-Gen Wang, Yanyan Han, Qing-Yu He. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 2012, 16(5):284-287.</code></pre>
<pre><code>## 
## Attaching package: &#39;clusterProfiler&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:AnnotationDbi&#39;:
## 
##     select</code></pre>
<pre><code>## The following object is masked from &#39;package:DelayedArray&#39;:
## 
##     simplify</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     slice</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     rename</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     filter</code></pre>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="gene-expression-based-enrichment-analysis.html#cb90-1"></a>ans.go &lt;-<span class="st"> </span><span class="kw">enrichGO</span>(<span class="dt">gene =</span> deGenes, <span class="dt">ont =</span> <span class="st">&quot;BP&quot;</span>,</span>
<span id="cb90-2"><a href="gene-expression-based-enrichment-analysis.html#cb90-2"></a>                   <span class="dt">OrgDb =</span><span class="st">&quot;org.Hs.eg.db&quot;</span>,</span>
<span id="cb90-3"><a href="gene-expression-based-enrichment-analysis.html#cb90-3"></a>                   <span class="dt">readable=</span><span class="ot">TRUE</span>,</span>
<span id="cb90-4"><a href="gene-expression-based-enrichment-analysis.html#cb90-4"></a>                   <span class="dt">pvalueCutoff =</span> <span class="fl">0.05</span>)</span>
<span id="cb90-5"><a href="gene-expression-based-enrichment-analysis.html#cb90-5"></a>tab.go &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(ans.go)</span>
<span id="cb90-6"><a href="gene-expression-based-enrichment-analysis.html#cb90-6"></a>tab.go&lt;-<span class="st"> </span><span class="kw">subset</span>(tab.go, Count<span class="op">&gt;</span><span class="dv">5</span>)</span>
<span id="cb90-7"><a href="gene-expression-based-enrichment-analysis.html#cb90-7"></a>tab.go[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre><code>##                    ID                          Description GeneRatio   BgRatio
## GO:0007411 GO:0007411                        axon guidance    26/547 276/18670
## GO:0097485 GO:0097485           neuron projection guidance    26/547 277/18670
## GO:0030198 GO:0030198    extracellular matrix organization    30/547 368/18670
## GO:0043062 GO:0043062 extracellular structure organization    30/547 369/18670
## GO:0045444 GO:0045444             fat cell differentiation    22/547 223/18670
##                  pvalue     p.adjust
## GO:0007411 1.776070e-07 0.0004337162
## GO:0097485 1.907283e-07 0.0004337162
## GO:0030198 4.818478e-07 0.0005802757
## GO:0043062 5.103568e-07 0.0005802757
## GO:0045444 7.186823e-07 0.0006537134</code></pre>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="gene-expression-based-enrichment-analysis.html#cb92-1"></a>ans.kegg &lt;-<span class="st"> </span><span class="kw">enrichKEGG</span>(<span class="dt">gene =</span> deGenes,</span>
<span id="cb92-2"><a href="gene-expression-based-enrichment-analysis.html#cb92-2"></a>                       <span class="dt">organism =</span> <span class="st">&#39;hsa&#39;</span>,</span>
<span id="cb92-3"><a href="gene-expression-based-enrichment-analysis.html#cb92-3"></a>                       <span class="dt">pvalueCutoff =</span> <span class="fl">0.05</span>)</span></code></pre></div>
<pre><code>## Reading KEGG annotation online:
## 
## Reading KEGG annotation online:</code></pre>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="gene-expression-based-enrichment-analysis.html#cb94-1"></a>tab.kegg &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(ans.kegg)</span>
<span id="cb94-2"><a href="gene-expression-based-enrichment-analysis.html#cb94-2"></a>tab.kegg&lt;-<span class="st"> </span><span class="kw">subset</span>(tab.kegg, Count<span class="op">&gt;</span><span class="dv">5</span>)</span>
<span id="cb94-3"><a href="gene-expression-based-enrichment-analysis.html#cb94-3"></a>tab.kegg[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre><code>##                ID                                          Description
## hsa04923 hsa04923                Regulation of lipolysis in adipocytes
## hsa04213 hsa04213      Longevity regulating pathway - multiple species
## hsa04360 hsa04360                                        Axon guidance
## hsa04933 hsa04933 AGE-RAGE signaling pathway in diabetic complications
## hsa04068 hsa04068                               FoxO signaling pathway
##          GeneRatio  BgRatio       pvalue    p.adjust
## hsa04923    10/286  57/8087 2.587664e-05 0.007074392
## hsa04213    10/286  62/8087 5.510596e-05 0.007074392
## hsa04360    18/286 182/8087 7.318336e-05 0.007074392
## hsa04933    12/286 100/8087 1.975759e-04 0.011981969
## hsa04068    14/286 131/8087 2.065857e-04 0.011981969</code></pre>
<p>For external databases such as DisGeNET or MSigDB collections, databases must be imported into R. This is how <a href="https://www.disgenet.org/">DisGeNET</a> is imported into R</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="gene-expression-based-enrichment-analysis.html#cb96-1"></a>gda &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="st">&quot;data/curated_gene_disease_associations.tsv.gz&quot;</span>)</span>
<span id="cb96-2"><a href="gene-expression-based-enrichment-analysis.html#cb96-2"></a>disease2gene &lt;-<span class="st"> </span>gda[, <span class="kw">c</span>(<span class="st">&quot;diseaseId&quot;</span>, <span class="st">&quot;geneId&quot;</span>)]</span>
<span id="cb96-3"><a href="gene-expression-based-enrichment-analysis.html#cb96-3"></a>disease2name &lt;-<span class="st"> </span>gda[, <span class="kw">c</span>(<span class="st">&quot;diseaseId&quot;</span>, <span class="st">&quot;diseaseName&quot;</span>)]</span></code></pre></div>
<p>Then, the enrichment is perform using <code>enricher</code> function:</p>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="gene-expression-based-enrichment-analysis.html#cb97-1"></a>ans.dis &lt;-<span class="st"> </span><span class="kw">enricher</span>(deGenes, <span class="dt">TERM2GENE=</span>disease2gene,</span>
<span id="cb97-2"><a href="gene-expression-based-enrichment-analysis.html#cb97-2"></a>                    <span class="dt">TERM2NAME=</span>disease2name)</span>
<span id="cb97-3"><a href="gene-expression-based-enrichment-analysis.html#cb97-3"></a>tab.dis &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(ans.dis)</span>
<span id="cb97-4"><a href="gene-expression-based-enrichment-analysis.html#cb97-4"></a>tab.dis&lt;-<span class="st"> </span><span class="kw">subset</span>(tab.dis, Count<span class="op">&gt;</span><span class="dv">5</span>)</span>
<span id="cb97-5"><a href="gene-expression-based-enrichment-analysis.html#cb97-5"></a>tab.dis[,<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre><code>##                ID         Description GeneRatio  BgRatio       pvalue
## C0014175 C0014175       Endometriosis    25/334 156/8948 5.455196e-10
## C0151744 C0151744 Myocardial Ischemia    21/334 171/8948 1.415760e-06
##              p.adjust
## C0014175 8.319175e-07
## C0151744 1.079517e-03</code></pre>
<p>This is the R code for transcription factors which is one of the sub-collection of C3 available at <a href="http://software.broadinstitute.org/gsea/msigdb">MSigDb</a> for
additional gene set collections.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="gene-expression-based-enrichment-analysis.html#cb99-1"></a>c3.tf &lt;-<span class="st"> </span><span class="kw">read.gmt</span>(<span class="st">&quot;c:/Juan/CREAL/HELIX/pathways/GSEA/c3.tft.v6.2.entrez.gmt&quot;</span>)</span>
<span id="cb99-2"><a href="gene-expression-based-enrichment-analysis.html#cb99-2"></a></span>
<span id="cb99-3"><a href="gene-expression-based-enrichment-analysis.html#cb99-3"></a>ans.tf &lt;-<span class="st"> </span><span class="kw">enricher</span>(deGenes, <span class="dt">TERM2GENE=</span>c3.tf)</span>
<span id="cb99-4"><a href="gene-expression-based-enrichment-analysis.html#cb99-4"></a>tab.tf &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(ans.tf)</span>
<span id="cb99-5"><a href="gene-expression-based-enrichment-analysis.html#cb99-5"></a>tab.tf&lt;-<span class="st"> </span><span class="kw">subset</span>(tab.tf, Count<span class="op">&gt;</span><span class="dv">5</span>)</span>
<span id="cb99-6"><a href="gene-expression-based-enrichment-analysis.html#cb99-6"></a>tab.tf[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</span></code></pre></div>
<pre><code>##                ID Description GeneRatio   BgRatio       pvalue
## HNF3_Q6   HNF3_Q6     HNF3_Q6    20/459 189/12774 1.520418e-05
## SRF_Q6     SRF_Q6      SRF_Q6    23/459 241/12774 1.947737e-05
## CEBPB_02 CEBPB_02    CEBPB_02    24/459 258/12774 1.972855e-05
## CEBP_Q2   CEBP_Q2     CEBP_Q2    22/459 234/12774 3.707975e-05
## OCT1_05   OCT1_05     OCT1_05    23/459 254/12774 4.481583e-05</code></pre>
</div>
<div id="visualization" class="section level2" number="9.4">
<h2><span class="header-section-number">9.4</span> Visualization</h2>
<p>All analyses performed with <code>clusterProfiler</code> can be visualize by different plots:</p>
<div class="sourceCode" id="cb101"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb101-1"><a href="gene-expression-based-enrichment-analysis.html#cb101-1"></a><span class="kw">library</span>(enrichplot)</span>
<span id="cb101-2"><a href="gene-expression-based-enrichment-analysis.html#cb101-2"></a>p1 &lt;-<span class="st"> </span><span class="kw">barplot</span>(ans.dis, <span class="dt">showCategory=</span><span class="dv">10</span>)</span>
<span id="cb101-3"><a href="gene-expression-based-enrichment-analysis.html#cb101-3"></a>p1</span></code></pre></div>
<p><img src="_main_files/figure-html/plot-1.png" width="672" /></p>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="gene-expression-based-enrichment-analysis.html#cb102-1"></a>p2 &lt;-<span class="st"> </span><span class="kw">dotplot</span>(ans.kegg, <span class="dt">showCategory=</span><span class="dv">20</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;KEGG&quot;</span>)</span>
<span id="cb102-2"><a href="gene-expression-based-enrichment-analysis.html#cb102-2"></a>p3 &lt;-<span class="st"> </span><span class="kw">dotplot</span>(ans.dis, <span class="dt">showCategory=</span><span class="dv">20</span>) <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;Disease&quot;</span>)</span>
<span id="cb102-3"><a href="gene-expression-based-enrichment-analysis.html#cb102-3"></a><span class="kw">plot_grid</span>(p2, p3, <span class="dt">nrow=</span><span class="dv">2</span>)</span></code></pre></div>
<p><img src="_main_files/figure-html/plot2-1.png" width="672" /></p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="gene-expression-based-enrichment-analysis.html#cb103-1"></a>p4 &lt;-<span class="st"> </span><span class="kw">upsetplot</span>(ans.dis)</span>
<span id="cb103-2"><a href="gene-expression-based-enrichment-analysis.html#cb103-2"></a>p4</span></code></pre></div>
<p><img src="_main_files/figure-html/plot3-1.png" width="672" /></p>
<div class="sourceCode" id="cb104"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="gene-expression-based-enrichment-analysis.html#cb104-1"></a>p5 &lt;-<span class="st"> </span><span class="kw">emapplot</span>(ans.kegg)</span>
<span id="cb104-2"><a href="gene-expression-based-enrichment-analysis.html#cb104-2"></a>p5</span></code></pre></div>
<p><img src="_main_files/figure-html/plot4-1.png" width="672" /></p>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="gene-expression-based-enrichment-analysis.html#cb105-1"></a>cowplot<span class="op">::</span><span class="kw">plot_grid</span>(p1, p3, p5, <span class="dt">ncol=</span><span class="dv">2</span>, <span class="dt">labels=</span>LETTERS[<span class="dv">1</span><span class="op">:</span><span class="dv">3</span>])</span></code></pre></div>
<p><img src="_main_files/figure-html/plot5-1.png" width="960" /></p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="introduction.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="genomic-region-enrichment-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/isglobal-brge/Master_Bioinformatics/tree/master/docs05-Enrichment_analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
