<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9 Exercises | Master in Bioinformatics and Omic Data Analysis</title>
  <meta name="description" content="9 Exercises | Master in Bioinformatics and Omic Data Analysis" />
  <meta name="generator" content="bookdown 0.20 and GitBook 2.6.7" />

  <meta property="og:title" content="9 Exercises | Master in Bioinformatics and Omic Data Analysis" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9 Exercises | Master in Bioinformatics and Omic Data Analysis" />
  
  
  

<meta name="author" content="Juan R González" />


<meta name="date" content="2021-01-26" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="references.html"/>

<script src="libs/header-attrs-2.3/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="bioconductor.html"><a href="bioconductor.html"><i class="fa fa-check"></i><b>2</b> Bioconductor</a>
<ul>
<li class="chapter" data-level="2.1" data-path="bioconductor.html"><a href="bioconductor.html#bioconductors-overview"><i class="fa fa-check"></i><b>2.1</b> Bioconductor’s overview</a></li>
<li class="chapter" data-level="2.2" data-path="bioconductor.html"><a href="bioconductor.html#bioconductor-infrastructures"><i class="fa fa-check"></i><b>2.2</b> Bioconductor infrastructures</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="bioconductor.html"><a href="bioconductor.html#snpmatrix"><i class="fa fa-check"></i><b>2.2.1</b> <code>snpMatrix</code></a></li>
<li class="chapter" data-level="2.2.2" data-path="bioconductor.html"><a href="bioconductor.html#expressionset"><i class="fa fa-check"></i><b>2.2.2</b> <code>ExpressionSet</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="bioconductor.html"><a href="bioconductor.html#summarizedexperiment"><i class="fa fa-check"></i><b>2.2.3</b> <code>SummarizedExperiment</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="bioconductor.html"><a href="bioconductor.html#genomic-ranges-granges"><i class="fa fa-check"></i><b>2.2.4</b> <code>Genomic Ranges (GRanges)</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="bioconductor.html"><a href="bioconductor.html#rangedsummarizedexperiment"><i class="fa fa-check"></i><b>2.2.5</b> <code>RangedSummarizedExperiment</code></a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="bioconductor.html"><a href="bioconductor.html#annotation"><i class="fa fa-check"></i><b>2.3</b> Annotation</a></li>
<li class="chapter" data-level="2.4" data-path="bioconductor.html"><a href="bioconductor.html#annotate-package"><i class="fa fa-check"></i><b>2.4</b> <code>annotate</code> package</a></li>
<li class="chapter" data-level="2.5" data-path="bioconductor.html"><a href="bioconductor.html#biomart"><i class="fa fa-check"></i><b>2.5</b> BioMart</a></li>
<li class="chapter" data-level="2.6" data-path="bioconductor.html"><a href="bioconductor.html#annotate-snps"><i class="fa fa-check"></i><b>2.6</b> Annotate SNPs</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="single-association-analysis.html"><a href="single-association-analysis.html"><i class="fa fa-check"></i><b>3</b> Single association analysis</a>
<ul>
<li class="chapter" data-level="3.1" data-path="single-association-analysis.html"><a href="single-association-analysis.html#descriptive-analysis"><i class="fa fa-check"></i><b>3.1</b> Descriptive analysis</a></li>
<li class="chapter" data-level="3.2" data-path="single-association-analysis.html"><a href="single-association-analysis.html#hardy-weinberg-equilibrium"><i class="fa fa-check"></i><b>3.2</b> Hardy-Weinberg equilibrium</a></li>
<li class="chapter" data-level="3.3" data-path="single-association-analysis.html"><a href="single-association-analysis.html#single-snp-association-analysis"><i class="fa fa-check"></i><b>3.3</b> Single SNP association analysis</a></li>
<li class="chapter" data-level="3.4" data-path="single-association-analysis.html"><a href="single-association-analysis.html#gene-environment-and-gene-gene-interactions"><i class="fa fa-check"></i><b>3.4</b> Gene-environment and gene-gene interactions</a></li>
<li class="chapter" data-level="3.5" data-path="single-association-analysis.html"><a href="single-association-analysis.html#haplotype-analysis"><i class="fa fa-check"></i><b>3.5</b> Haplotype analysis</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="single-association-analysis.html"><a href="single-association-analysis.html#haplotype-estimation"><i class="fa fa-check"></i><b>3.5.1</b> Haplotype estimation</a></li>
<li class="chapter" data-level="3.5.2" data-path="single-association-analysis.html"><a href="single-association-analysis.html#haplotype-association"><i class="fa fa-check"></i><b>3.5.2</b> Haplotype association</a></li>
<li class="chapter" data-level="3.5.3" data-path="single-association-analysis.html"><a href="single-association-analysis.html#sliding-window-approach"><i class="fa fa-check"></i><b>3.5.3</b> Sliding window approach</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="single-association-analysis.html"><a href="single-association-analysis.html#genetic-score"><i class="fa fa-check"></i><b>3.6</b> Genetic score</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html"><i class="fa fa-check"></i><b>4</b> Genome-wide association studies</a>
<ul>
<li class="chapter" data-level="4.1" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#quality-control-of-snps"><i class="fa fa-check"></i><b>4.1</b> Quality control of SNPs</a></li>
<li class="chapter" data-level="4.2" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#quality-control-of-individuals"><i class="fa fa-check"></i><b>4.2</b> Quality control of individuals</a></li>
<li class="chapter" data-level="4.3" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#population-ancestry"><i class="fa fa-check"></i><b>4.3</b> Population ancestry</a></li>
<li class="chapter" data-level="4.4" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#genome-wide-association-analysis"><i class="fa fa-check"></i><b>4.4</b> Genome-wide association analysis</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="microarray-data-analysis.html"><a href="microarray-data-analysis.html"><i class="fa fa-check"></i><b>5</b> Microarray data analysis</a></li>
<li class="chapter" data-level="6" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html"><i class="fa fa-check"></i><b>6</b> RNAseq data analysis</a>
<ul>
<li class="chapter" data-level="6.1" data-path="index.html"><a href="index.html#introduction"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#within-sample-normalization-of-read-counts"><i class="fa fa-check"></i><b>6.2</b> Within sample normalization of read counts</a></li>
<li class="chapter" data-level="6.3" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#exploratory-analysis-of-the-read-counts-table"><i class="fa fa-check"></i><b>6.3</b> Exploratory analysis of the read counts table</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#clustering"><i class="fa fa-check"></i><b>6.3.1</b> Clustering</a></li>
<li class="chapter" data-level="6.3.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#pca"><i class="fa fa-check"></i><b>6.3.2</b> PCA</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.4</b> Differential expression analysis</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#deseq2"><i class="fa fa-check"></i><b>6.4.1</b> DESeq2</a></li>
<li class="chapter" data-level="6.4.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#voom"><i class="fa fa-check"></i><b>6.4.2</b> voom</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#accounting-for-additional-sources-of-variation"><i class="fa fa-check"></i><b>6.5</b> Accounting for additional sources of variation</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#accounting-for-covariates-using-deseq2"><i class="fa fa-check"></i><b>6.5.1</b> Accounting for covariates using <code>DESeq2</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#accounting-for-estimated-covariates-using-sva-and-ruvseq"><i class="fa fa-check"></i><b>6.5.2</b> Accounting for estimated covariates using <code>sva</code> and <code>RUVSeq</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html"><i class="fa fa-check"></i><b>7</b> Enrichment Analysis</a>
<ul>
<li class="chapter" data-level="7.1" data-path="index.html"><a href="index.html#introduction"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#gene-expression-based-enrichment-analysis"><i class="fa fa-check"></i><b>7.2</b> Gene expression-based enrichment analysis</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#a-primer-on-terminology-existing-methods-statistical-theory"><i class="fa fa-check"></i><b>7.2.1</b> A primer on terminology, existing methods &amp; statistical theory</a></li>
<li class="chapter" data-level="7.2.2" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#basic-analysis-using-gostats"><i class="fa fa-check"></i><b>7.2.2</b> Basic analysis using <code>GOstats</code></a></li>
<li class="chapter" data-level="7.2.3" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#analysis-using-clusterprofiler"><i class="fa fa-check"></i><b>7.2.3</b> Analysis using <code>clusterProfiler</code></a></li>
<li class="chapter" data-level="7.2.4" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#visualization"><i class="fa fa-check"></i><b>7.2.4</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#genomic-region-enrichment-analysis"><i class="fa fa-check"></i><b>7.3</b> Genomic region enrichment analysis</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>8</b> References</a></li>
<li class="chapter" data-level="9" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>9</b> Exercises</a>
<ul>
<li class="chapter" data-level="9.1" data-path="exercises.html"><a href="exercises.html#BioCEx"><i class="fa fa-check"></i><b>9.1</b> Bioconductor</a></li>
<li class="chapter" data-level="9.2" data-path="exercises.html"><a href="exercises.html#genetic-association-studies"><i class="fa fa-check"></i><b>9.2</b> Genetic association studies</a></li>
<li class="chapter" data-level="9.3" data-path="exercises.html"><a href="exercises.html#gwas"><i class="fa fa-check"></i><b>9.3</b> GWAS</a></li>
<li class="chapter" data-level="9.4" data-path="exercises.html"><a href="exercises.html#Microarray-exerc"><i class="fa fa-check"></i><b>9.4</b> Microarray data analysis</a></li>
<li class="chapter" data-level="9.5" data-path="exercises.html"><a href="exercises.html#RNAseq-exerc"><i class="fa fa-check"></i><b>9.5</b> RNAseq data analysis</a></li>
<li class="chapter" data-level="9.6" data-path="exercises.html"><a href="exercises.html#enrichment-exerc"><i class="fa fa-check"></i><b>9.6</b> Enrichment Analysis</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Master in Bioinformatics and Omic Data Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="exercises" class="section level1" number="9">
<h1><span class="header-section-number">9</span> Exercises</h1>
<p>Here we propose several exercises whose answers will have to be deliver through the Moodle. The data are in a folder called <code>data_for_exercises</code>. The answers will also be available in the Moodle site.</p>
<div id="BioCEx" class="section level2" number="9.1">
<h2><span class="header-section-number">9.1</span> Bioconductor</h2>
<blockquote>
<p><strong>EXERCISE 1</strong>: Read PLINK data (‘obesity.bed’, ‘obesity.fam’, ‘obesity.bim’) that are available <a href="https://github.com/isglobal-brge/brgedata/tree/master/inst/extdata">here</a></p>
<ol style="list-style-type: decimal">
<li>How many samples are genotyped?</li>
<li>How many SNPs are in the chromosome 10?</li>
<li>Get the SNP names available in the region chr2:134500-2300000.</li>
<li>In which chromosome is located the SNP rs1527354?</li>
<li>Get the number of individuals of each genoytpe corresponding to the SNP rs7859525. Which the homozygous normal genotype?</li>
</ol>
</blockquote>
<blockquote>
<p><strong>EXERCISE 2:</strong> Install the Bioconductor package called <code>airway</code> and load the data <code>airway</code> which is a <code>SummarizedExperiment</code> by executing:</p>
<div class="sourceCode" id="cb466"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb466-1"><a href="exercises.html#cb466-1"></a><span class="kw">library</span>(airway)</span>
<span id="cb466-2"><a href="exercises.html#cb466-2"></a><span class="kw">data</span>(airway)</span></code></pre></div>
<ol style="list-style-type: decimal">
<li>Print the object {} and interpret the output.</li>
<li>Get the table of counts (use {} data) and the variables of the experiment (use {} function).</li>
<li>Verify that {} objects check the dimension of the tables (think about the dimensions that should match).</li>
<li>Compare the number of counts of the gene ENSG00000000419 between treated and untreated cells (variable ‘dex’) using {} function.</li>
</ol>
</blockquote>
<blockquote>
<p><strong>EXERCISE 3:</strong> <a href="https://jhubiostatistics.shinyapps.io/recount/">Recount2</a> provides data for different RNA-seq experiments. These includes data from GTEx or TCGA projects. We have downloaded a subset of data corresponding to breast cancer and created a variable called <code>er</code> which encodes the estrogen receptor status (Negative and Positive). The <code>SummarizedExperiment</code> object is called <code>breast</code> and is available in the file ‘breast_tcga.Rdata’. Load the data into R and answer the next questions</p>
<ol style="list-style-type: decimal">
<li>How many samples are in the <code>SummarizedExperiment</code> object?</li>
<li>And how many genes?</li>
<li>Which is the number of samples with positive estrogen receptor status (variable <code>er</code>)?</li>
<li>Subset the individuals having Negative strogen receptor status and draw a boxplot of the first gene.</li>
<li>Create a <code>SummarizedExperiment</code> object of the genomic region chr6:151.2-151.8Mb. How many genes are in that region? How many of them are annotated? That is, how many of them have a gene symbol name? (HINT: use <code>rowRanges()</code> function and remember that <code>mcols()</code> function is used to get acccess to columns in a <code>GRanges</code> object)</li>
</ol>
</blockquote>
<blockquote>
<p><strong>EXERCISE 4</strong>: The file <code>snpsGWAS.Rdata</code> contains information about 28 SNPs that were called as a significantly associated with a given disease after performing a GWAS study.</p>
<ol style="list-style-type: decimal">
<li>Load this data into R.</li>
<li>Create a data.frame with the name of the SNP, its chromosome, its position and the allele name.</li>
<li>Add another column to this data.frame with the minor allele frequency (MAF) estimated from 1000 genomes (HINT: use <code>listAttributes</code> to retrieve the name of the attribute we are interested in).</li>
<li>Annotate the variants and get the gene symbol</li>
<li>How many SNPs are located in promoter regions?</li>
</ol>
</blockquote>
</div>
<div id="genetic-association-studies" class="section level2" number="9.2">
<h2><span class="header-section-number">9.2</span> Genetic association studies</h2>
<blockquote>
<p><strong>EXERCISE 5</strong>: Researchers are interested in assessing possible association between candidate SNPs and the response to treatment in patients diagnosed with major depression (file ‘DM.txt’ - NOTE: check how alleles are separated). The file also includes clinical information about other covariates of interest such as:</p>
<ul>
<li>HDRS: Hamilton Depression Rating Scale (continuous variable)</li>
<li>PSICOT: Was the patient psychotic? (No and Yes)</li>
<li>MELANCOL: Was the patient melancholic? (No and Yes)</li>
<li>EPD_PREV: Number of previous episodes of depression (continuous variable)</li>
<li>RESP: Response to treatment (outcome)</li>
</ul>
<p>Answer the following questions:</p>
<ol style="list-style-type: decimal">
<li>Is it necessary to check HWE hypothesis for this example?</li>
<li>Is there any SNP associated with the response to the treatment? If so, write a sentence interpreting the results of that association (only for one SNP)</li>
<li>Does the result change after adjusting for other clinical covariates?</li>
<li>Create a plot with the p-values only for dominant, recessive and additive models</li>
<li>Compute the p-values using MAX-statistic for all SNPs</li>
</ol>
</blockquote>
<blockquote>
<p><strong>EXERCISE 6</strong>: Researchers are now interested in assessing the effect of haplotypes on the response to treatment. Answer the following questions:</p>
<ol style="list-style-type: decimal">
<li>Which is the combination of SNPs (e.g. number of consecutive haplotypes) that is more associated with the response to treatment? (NOTE1: Try only haplotypes of length 2 up to 4)</li>
<li>Which is the most frequent haplotype?</li>
<li>Compute the OR of association and its 95% confidence intervals between the estimate haplotypes and the response to treatment</li>
</ol>
</blockquote>
<blockquote>
<p><strong>EXERCISE 7</strong>: Researchers are now interested in creating a genetic score to predict the response to the treatment</p>
<ol style="list-style-type: decimal">
<li>Perform all the required steps to get such score and evaluate its performance using a ROC curve</li>
</ol>
</blockquote>
</div>
<div id="gwas" class="section level2" number="9.3">
<h2><span class="header-section-number">9.3</span> GWAS</h2>
<blockquote>
<p><strong>EXERCISE 8</strong>: Researchers are interested in detecting new SNPs associated with BMI (body mass index). To do so, they performed a GWAS using DNA information about 425 individuals. Genotype information is available in plink format (files ‘coronary.bed’, ‘coronary.bim’, ‘coronary.fam’) while phenotypic information can be found in the file ‘coronary.txt’.</p>
<ol style="list-style-type: decimal">
<li>Read genotypes using <code>snpStats</code> library</li>
<li>Verify that both data sets are in the same order (NOTE: ‘id’ variable in the file ‘coronary.txt’ must be used since it corresponds to the unique patient number)</li>
<li>Remove those SNPs that do not pass QC</li>
<li>Perform QC filtering of individuals</li>
<li>Assess association between BMI’ and the SNPs (NOTE: remember that you are analyzing a <strong>quantitative trait</strong>)</li>
<li>Calculate ‘lambda’ and create a Q-Q plot to assess population stratification</li>
<li>Assess association between BMI’ and the SNPs adjusting for population stratification (variables ‘ev3’ and ‘ev4’ in the file ‘coronary.txt’) (NOTE: remember that you are analyzing a <strong>quantitative trait</strong>). Are there differences with crude analysis?</li>
<li>Create a Manhattan plot</li>
</ol>
</blockquote>
</div>
<div id="Microarray-exerc" class="section level2" number="9.4">
<h2><span class="header-section-number">9.4</span> Microarray data analysis</h2>
<blockquote>
<p><strong>EXERCISE 9</strong>: To be supplied</p>
</blockquote>
</div>
<div id="RNAseq-exerc" class="section level2" number="9.5">
<h2><span class="header-section-number">9.5</span> RNAseq data analysis</h2>
<blockquote>
<p><strong>EXERICSE 10</strong>: Download the data SRP029880 from recount2: <a href="http://duffel.rail.bio/recount/v2/SRP029880/rse_gene.Rdata" class="uri">http://duffel.rail.bio/recount/v2/SRP029880/rse_gene.Rdata</a>. This file contains a <code>RangedSummarizedExperiment</code> object. Do the following tasks:</p>
<ol start="0" style="list-style-type: decimal">
<li>Get the <code>counts</code> and <code>colData</code> tables.</li>
<li>Normalize the counts using CPM, RPKM, TPM and TMM methods.</li>
<li>Create the RLE and PCA plots for each method as well as for the raw data (do not forget to log-transform the data).</li>
<li>Using TPM normalized data:</li>
</ol>
<ul>
<li>Plot a heatmap of the top 500 most variable genes. Compare with the heatmap obtained using the 100 most variable genes.</li>
<li>Re-do the heatmaps setting the scale argument to <code>none</code>, and <code>column</code>. Compare the results with <code>scale = 'row'</code>.</li>
</ul>
<ol start="4" style="list-style-type: decimal">
<li>Create a new variable called <strong>group</strong> by renaming the variable <strong>title</strong> as: <code>normal</code>, <code>primary</code> and <code>metastasis</code>.</li>
<li>Perform a differential expression analysis using the variable <code>group</code> and get the list of differentially expressed genes comparing <code>metastasis vs normal</code> and <code>primary vs normal</code>.</li>
<li>Create two MA and volcano plots to describe the previous comparisons.</li>
<li>Examine the distribution of counts for a single gene (most associated with <code>metastasis</code>). Hint: use <code>DESeq2::plotCounts ()</code></li>
<li>Do the same plot for the top-10 genes (this tries to mimic your supplementary material in your paper!)</li>
<li>Investigate whether the analysis should be adjusted for other covariates.</li>
<li>Repeat differential expression analysis accounting for unwanted variability using <code>sva</code> package.</li>
</ol>
</blockquote>
</div>
<div id="enrichment-exerc" class="section level2" number="9.6">
<h2><span class="header-section-number">9.6</span> Enrichment Analysis</h2>
<blockquote>
<p><strong>EXERCISE 11</strong>: Library <code>tweeDEseqCountData</code> contains data corresponding to an RNA-seq experiment described in Pickrell et al. (2010). Data correspond to lymphoblastoid cell lines about 69 non-related Nigerian individuals. This information as well as phenotypic data is available as an object of class <code>eSet</code> called <code>pickrell.eset</code> that can be loaded after typing:</p>
<pre><code>data(pickrell)</code></pre>
<ol style="list-style-type: decimal">
<li><p>Use limma-voom method to detect those genes that are differentially expressed (DE) between males and females (variale <code>gender</code>).</p></li>
<li><p>Select those genes that are DE at 5% FDR with a minimum fold-change of 1.5</p></li>
<li><p>Create a gene set of sex-expecific genes by</p></li>
</ol>
<pre><code>geneUniverse &lt;- featureNames(pickrell.eset)
geneSex &lt;- unique(intersect(geneUniverse,
c(msYgenes, XiEgenes)))</code></pre>
<p>and test whether the list of DE genes is enriched in that gene set.</p>
</blockquote>
<blockquote>
<p><strong>EXERCISE 12</strong>: Use the data in exercise 10 to obtain a list of genes that are differentially expressed between <code>metastasis vs normal</code> samples (you can run the R script that is available in the Moodle - ‘Answer exercises’)</p>
<ol style="list-style-type: decimal">
<li><p>Perform an enrichment analysis of GO, KEGG and DisGeNET databases using <code>clusterProfiler</code>.</p></li>
<li><p>Investigate whether the DE genes are enriched in protein-coding genes.</p></li>
</ol>
</blockquote>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="references.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/isglobal-brge/Master_Bioinformatics/tree/master/docs99-Exercises.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
