<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 Single association analysis | Master in Bioinformatics and Omic Data Analysis</title>
  <meta name="description" content="3 Single association analysis | Master in Bioinformatics and Omic Data Analysis" />
  <meta name="generator" content="bookdown 0.23 and GitBook 2.6.7" />

  <meta property="og:title" content="3 Single association analysis | Master in Bioinformatics and Omic Data Analysis" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 Single association analysis | Master in Bioinformatics and Omic Data Analysis" />
  
  
  

<meta name="author" content="Juan R González" />


<meta name="date" content="2021-11-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="bioconductor.html"/>
<link rel="next" href="genome-wide-association-studies.html"/>
<script src="libs/header-attrs-2.10/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a></li>
<li class="chapter" data-level="2" data-path="bioconductor.html"><a href="bioconductor.html"><i class="fa fa-check"></i><b>2</b> Bioconductor</a>
<ul>
<li class="chapter" data-level="2.1" data-path="bioconductor.html"><a href="bioconductor.html#bioconductors-overview"><i class="fa fa-check"></i><b>2.1</b> Bioconductor’s overview</a></li>
<li class="chapter" data-level="2.2" data-path="bioconductor.html"><a href="bioconductor.html#bioconductor-infrastructures"><i class="fa fa-check"></i><b>2.2</b> Bioconductor infrastructures</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="bioconductor.html"><a href="bioconductor.html#snpmatrix"><i class="fa fa-check"></i><b>2.2.1</b> <code>snpMatrix</code></a></li>
<li class="chapter" data-level="2.2.2" data-path="bioconductor.html"><a href="bioconductor.html#expressionset"><i class="fa fa-check"></i><b>2.2.2</b> <code>ExpressionSet</code></a></li>
<li class="chapter" data-level="2.2.3" data-path="bioconductor.html"><a href="bioconductor.html#summarizedexperiment"><i class="fa fa-check"></i><b>2.2.3</b> <code>SummarizedExperiment</code></a></li>
<li class="chapter" data-level="2.2.4" data-path="bioconductor.html"><a href="bioconductor.html#genomic-ranges-granges"><i class="fa fa-check"></i><b>2.2.4</b> <code>Genomic Ranges (GRanges)</code></a></li>
<li class="chapter" data-level="2.2.5" data-path="bioconductor.html"><a href="bioconductor.html#rangedsummarizedexperiment"><i class="fa fa-check"></i><b>2.2.5</b> <code>RangedSummarizedExperiment</code></a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="bioconductor.html"><a href="bioconductor.html#annotation"><i class="fa fa-check"></i><b>2.3</b> Annotation</a></li>
<li class="chapter" data-level="2.4" data-path="bioconductor.html"><a href="bioconductor.html#annotate-package"><i class="fa fa-check"></i><b>2.4</b> <code>annotate</code> package</a></li>
<li class="chapter" data-level="2.5" data-path="bioconductor.html"><a href="bioconductor.html#biomart"><i class="fa fa-check"></i><b>2.5</b> BioMart</a></li>
<li class="chapter" data-level="2.6" data-path="bioconductor.html"><a href="bioconductor.html#annotate-snps"><i class="fa fa-check"></i><b>2.6</b> Annotate SNPs</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="single-association-analysis.html"><a href="single-association-analysis.html"><i class="fa fa-check"></i><b>3</b> Single association analysis</a>
<ul>
<li class="chapter" data-level="3.1" data-path="single-association-analysis.html"><a href="single-association-analysis.html#descriptive-analysis"><i class="fa fa-check"></i><b>3.1</b> Descriptive analysis</a></li>
<li class="chapter" data-level="3.2" data-path="single-association-analysis.html"><a href="single-association-analysis.html#hardy-weinberg-equilibrium"><i class="fa fa-check"></i><b>3.2</b> Hardy-Weinberg equilibrium</a></li>
<li class="chapter" data-level="3.3" data-path="single-association-analysis.html"><a href="single-association-analysis.html#single-snp-association-analysis"><i class="fa fa-check"></i><b>3.3</b> Single SNP association analysis</a></li>
<li class="chapter" data-level="3.4" data-path="single-association-analysis.html"><a href="single-association-analysis.html#gene-environment-and-gene-gene-interactions"><i class="fa fa-check"></i><b>3.4</b> Gene-environment and gene-gene interactions</a></li>
<li class="chapter" data-level="3.5" data-path="single-association-analysis.html"><a href="single-association-analysis.html#haplotype-analysis"><i class="fa fa-check"></i><b>3.5</b> Haplotype analysis</a>
<ul>
<li class="chapter" data-level="3.5.1" data-path="single-association-analysis.html"><a href="single-association-analysis.html#haplotype-estimation"><i class="fa fa-check"></i><b>3.5.1</b> Haplotype estimation</a></li>
<li class="chapter" data-level="3.5.2" data-path="single-association-analysis.html"><a href="single-association-analysis.html#haplotype-association"><i class="fa fa-check"></i><b>3.5.2</b> Haplotype association</a></li>
<li class="chapter" data-level="3.5.3" data-path="single-association-analysis.html"><a href="single-association-analysis.html#sliding-window-approach"><i class="fa fa-check"></i><b>3.5.3</b> Sliding window approach</a></li>
</ul></li>
<li class="chapter" data-level="3.6" data-path="single-association-analysis.html"><a href="single-association-analysis.html#genetic-score"><i class="fa fa-check"></i><b>3.6</b> Genetic score</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html"><i class="fa fa-check"></i><b>4</b> Genome-wide association studies</a>
<ul>
<li class="chapter" data-level="4.1" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#quality-control-of-snps"><i class="fa fa-check"></i><b>4.1</b> Quality control of SNPs</a></li>
<li class="chapter" data-level="4.2" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#quality-control-of-individuals"><i class="fa fa-check"></i><b>4.2</b> Quality control of individuals</a></li>
<li class="chapter" data-level="4.3" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#population-ancestry"><i class="fa fa-check"></i><b>4.3</b> Population ancestry</a></li>
<li class="chapter" data-level="4.4" data-path="genome-wide-association-studies.html"><a href="genome-wide-association-studies.html#genome-wide-association-analysis"><i class="fa fa-check"></i><b>4.4</b> Genome-wide association analysis</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="microarray-data-analysis.html"><a href="microarray-data-analysis.html"><i class="fa fa-check"></i><b>5</b> Microarray data analysis</a></li>
<li class="chapter" data-level="6" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html"><i class="fa fa-check"></i><b>6</b> RNAseq data analysis</a>
<ul>
<li class="chapter" data-level="6.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#introduction-1"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#within-sample-normalization-of-read-counts"><i class="fa fa-check"></i><b>6.2</b> Within sample normalization of read counts</a></li>
<li class="chapter" data-level="6.3" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#exploratory-analysis-of-the-read-counts-table"><i class="fa fa-check"></i><b>6.3</b> Exploratory analysis of the read counts table</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#clustering"><i class="fa fa-check"></i><b>6.3.1</b> Clustering</a></li>
<li class="chapter" data-level="6.3.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#pca"><i class="fa fa-check"></i><b>6.3.2</b> PCA</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#differential-expression-analysis"><i class="fa fa-check"></i><b>6.4</b> Differential expression analysis</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#deseq2"><i class="fa fa-check"></i><b>6.4.1</b> DESeq2</a></li>
<li class="chapter" data-level="6.4.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#voom"><i class="fa fa-check"></i><b>6.4.2</b> voom</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#accounting-for-additional-sources-of-variation"><i class="fa fa-check"></i><b>6.5</b> Accounting for additional sources of variation</a>
<ul>
<li class="chapter" data-level="6.5.1" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#accounting-for-covariates-using-deseq2"><i class="fa fa-check"></i><b>6.5.1</b> Accounting for covariates using <code>DESeq2</code></a></li>
<li class="chapter" data-level="6.5.2" data-path="rnaseq-data-analysis.html"><a href="rnaseq-data-analysis.html#accounting-for-estimated-covariates-using-sva-and-ruvseq"><i class="fa fa-check"></i><b>6.5.2</b> Accounting for estimated covariates using <code>sva</code> and <code>RUVSeq</code></a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html"><i class="fa fa-check"></i><b>7</b> Enrichment Analysis</a>
<ul>
<li class="chapter" data-level="7.1" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#introduction-2"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#gene-expression-based-enrichment-analysis"><i class="fa fa-check"></i><b>7.2</b> Gene expression-based enrichment analysis</a>
<ul>
<li class="chapter" data-level="7.2.1" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#a-primer-on-terminology-existing-methods-statistical-theory"><i class="fa fa-check"></i><b>7.2.1</b> A primer on terminology, existing methods &amp; statistical theory</a></li>
<li class="chapter" data-level="7.2.2" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#basic-analysis-using-gostats"><i class="fa fa-check"></i><b>7.2.2</b> Basic analysis using <code>GOstats</code></a></li>
<li class="chapter" data-level="7.2.3" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#analysis-using-clusterprofiler"><i class="fa fa-check"></i><b>7.2.3</b> Analysis using <code>clusterProfiler</code></a></li>
<li class="chapter" data-level="7.2.4" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#visualization"><i class="fa fa-check"></i><b>7.2.4</b> Visualization</a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="enrichment-analysis.html"><a href="enrichment-analysis.html#genomic-region-enrichment-analysis"><i class="fa fa-check"></i><b>7.3</b> Genomic region enrichment analysis</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>8</b> References</a></li>
<li class="chapter" data-level="9" data-path="exercises.html"><a href="exercises.html"><i class="fa fa-check"></i><b>9</b> Exercises</a>
<ul>
<li class="chapter" data-level="9.1" data-path="exercises.html"><a href="exercises.html#BioCEx"><i class="fa fa-check"></i><b>9.1</b> Bioconductor</a></li>
<li class="chapter" data-level="9.2" data-path="exercises.html"><a href="exercises.html#genetic-association-studies"><i class="fa fa-check"></i><b>9.2</b> Genetic association studies</a></li>
<li class="chapter" data-level="9.3" data-path="exercises.html"><a href="exercises.html#gwas"><i class="fa fa-check"></i><b>9.3</b> GWAS</a></li>
<li class="chapter" data-level="9.4" data-path="exercises.html"><a href="exercises.html#Microarray-exerc"><i class="fa fa-check"></i><b>9.4</b> Microarray data analysis</a></li>
<li class="chapter" data-level="9.5" data-path="exercises.html"><a href="exercises.html#RNAseq-exerc"><i class="fa fa-check"></i><b>9.5</b> RNAseq data analysis</a></li>
<li class="chapter" data-level="9.6" data-path="exercises.html"><a href="exercises.html#enrichment-exerc"><i class="fa fa-check"></i><b>9.6</b> Enrichment Analysis</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Master in Bioinformatics and Omic Data Analysis</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="single-association-analysis" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Single association analysis</h1>
<p>Genetic association studies aim to identify genetic variants that explain subject differences in qualitative or quantitative traits. Genetic variants may include SNPs, CNVs, mosaicisms or polymorphic inversions, which can be obtained from SNP array. In this chapter, we thoroughly discuss SNP association studies. We explain the different genetic models to be used in single SNP associations and how they are applied in a massive univariate testing when several SNPs are analyzed. Relating the simultaneous analysis of multiple SNPs, we discuss haplotype and genetic scores associations</p>
<div id="descriptive-analysis" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Descriptive analysis</h2>
<p>SNP data are typically available in text format or Excel spreadsheets which are easily uploaded in <code>R</code> as a data frame. Here, as an illustrative example, we are analyzing a dataset containing epidemiological information and 51 SNPs from a case-control study on asthma. The data is loaded into the R session by</p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb148-1"><a href="single-association-analysis.html#cb148-1" aria-hidden="true" tabindex="-1"></a>asthma <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">&quot;data/asthma.txt&quot;</span>)</span>
<span id="cb148-2"><a href="single-association-analysis.html#cb148-2" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(asthma, <span class="at">list.len=</span><span class="dv">9</span>)</span></code></pre></div>
<pre><code>&#39;data.frame&#39;:   1578 obs. of  57 variables:
 $ country    : chr  &quot;Germany&quot; &quot;Germany&quot; &quot;Germany&quot; &quot;Germany&quot; ...
 $ gender     : chr  &quot;Males&quot; &quot;Males&quot; &quot;Males&quot; &quot;Females&quot; ...
 $ age        : num  42.8 50.2 46.7 47.9 48.4 ...
 $ bmi        : num  20.1 24.7 27.7 33.3 25.2 ...
 $ smoke      : int  2 0 1 1 0 2 1 0 0 0 ...
 $ casecontrol: int  0 0 0 0 1 0 0 0 0 0 ...
 $ rs4490198  : chr  &quot;GG&quot; &quot;GG&quot; &quot;GG&quot; &quot;AG&quot; ...
 $ rs4849332  : chr  &quot;TT&quot; &quot;GT&quot; &quot;TT&quot; &quot;GT&quot; ...
 $ rs1367179  : chr  &quot;GC&quot; &quot;GC&quot; &quot;GC&quot; &quot;GG&quot; ...
  [list output truncated]</code></pre>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb150-1"><a href="single-association-analysis.html#cb150-1" aria-hidden="true" tabindex="-1"></a>asthma[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span></code></pre></div>
<pre><code>  country  gender      age      bmi smoke casecontrol rs4490198
1 Germany   Males 42.80630 20.14797     2           0        GG
2 Germany   Males 50.22861 24.69136     0           0        GG
3 Germany   Males 46.68857 27.73230     1           0        GG
4 Germany Females 47.86311 33.33187     1           0        AG
5 Germany Females 48.44079 25.23634     0           1        AG
  rs4849332
1        TT
2        GT
3        TT
4        GT
5        GG</code></pre>
<p>We observe that we have case-control status (0: control, 1: asthma) and another 4 variables encoding the country of origin, gender, age, body mass index (bmi) and smoking status (0: no smoker, 1: ex-smoker, 2: current smoker). There are 51 SNPs whose genotypes are given by the alleles names.</p>
<p>To start the analysis, we must indicate which columns of the dataset <code>asthma</code> contain the SNP data, using the <code>setupSNP</code> function. In our example, SNPs start from column 7 onwards, which we specify in argument <em>colSNPs</em></p>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb152-1"><a href="single-association-analysis.html#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SNPassoc)</span>
<span id="cb152-2"><a href="single-association-analysis.html#cb152-2" aria-hidden="true" tabindex="-1"></a>asthma.s <span class="ot">&lt;-</span> <span class="fu">setupSNP</span>(<span class="at">data=</span>asthma, <span class="at">colSNPs=</span><span class="dv">7</span><span class="sc">:</span><span class="fu">ncol</span>(asthma), <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>This is an alternative way of determining the columns containing the SNPs</p>
<div class="sourceCode" id="cb153"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb153-1"><a href="single-association-analysis.html#cb153-1" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">grep</span>(<span class="st">&quot;^rs&quot;</span>, <span class="fu">colnames</span>(asthma))</span>
<span id="cb153-2"><a href="single-association-analysis.html#cb153-2" aria-hidden="true" tabindex="-1"></a>asthma.s <span class="ot">&lt;-</span> <span class="fu">setupSNP</span>(<span class="at">data=</span>asthma, <span class="at">colSNPs=</span>idx, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>The argument <em>sep</em> indicates the character separating the alleles. The default value is ’‘/´´. In our case, there is no separating character, so that, we set <em>sep=""</em>. The argument <em>name.genotypes</em> can be used when genotypes are available in other formats, such as 0, 1, 2 or’‘norm´´,’‘het´´,’’mut´´. The purpose of the <code>setupSNP</code> function is to assign the class <em>snp</em> to the SNPs variables, to which <code>SNPassoc</code> methods will be applied. The function labels the most common genotype across subjects as the reference one. When numerous SNPs are available, the function can be parallelized through the argument <em>mc.cores</em> that indicates the number of processors to be used. We can verify that the SNP variables are given the new class <em>snp</em></p>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb154-1"><a href="single-association-analysis.html#cb154-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(asthma.s<span class="sc">$</span>rs1422993)</span></code></pre></div>
<pre><code>[1] G/G G/T G/G G/T G/T G/G
Genotypes: G/G G/T T/T
Alleles:  G T </code></pre>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb156-1"><a href="single-association-analysis.html#cb156-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(asthma.s<span class="sc">$</span>rs1422993)</span></code></pre></div>
<pre><code>[1] &quot;snp&quot;    &quot;factor&quot;</code></pre>
<p>and summarize their content with <code>summary</code></p>
<div class="sourceCode" id="cb158"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb158-1"><a href="single-association-analysis.html#cb158-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(asthma.s<span class="sc">$</span>rs1422993)</span></code></pre></div>
<pre><code>Genotypes: 
    frequency percentage
G/G       903  57.224335
G/T       570  36.121673
T/T       105   6.653992

Alleles: 
  frequency percentage
G      2376   75.28517
T       780   24.71483

HWE (p value): 0.250093 </code></pre>
<p>which shows the genotype and allele frequencies for a given SNP, testing for Hardy-Weinberg equilibrium (HWE). We can also visualize the results in a plot by</p>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb160-1"><a href="single-association-analysis.html#cb160-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(asthma.s<span class="sc">$</span>rs1422993)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plotsummarySNP"></span>
<img src="fig/plotsummarySNP-1.png" alt="SNP summary. Bar chart showing the basic information of a given SNP" width="672" />
<p class="caption">
Figure 3.1: SNP summary. Bar chart showing the basic information of a given SNP
</p>
</div>
<p>The argument <em>type</em> helps to get a pie chart</p>
<div class="sourceCode" id="cb161"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb161-1"><a href="single-association-analysis.html#cb161-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(asthma.s<span class="sc">$</span>rs1422993, <span class="at">type=</span>pie)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plotsummarySNP2"></span>
<img src="fig/plotsummarySNP2-1.png" alt="SNP summary. Pie chart showing the basic information of a given SNP" width="672" />
<p class="caption">
Figure 3.2: SNP summary. Pie chart showing the basic information of a given SNP
</p>
</div>
<p>The <em>summary</em> function can also be applied to the whole dataset</p>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb162-1"><a href="single-association-analysis.html#cb162-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(asthma.s, <span class="at">print=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>           alleles major.allele.freq HWE      missing (%)
rs4490198  A/G     59.2              0.174133  0.6       
rs4849332  G/T     61.8              0.522060  0.1       
rs1367179  G/C     81.4              0.738153  1.0       
rs11123242 C/T     81.7              0.932898  0.6       
rs13014858 G/A     58.3              0.351116  0.1       
rs1430094  G/A     66.9              0.305509  0.4       
rs1430093  C/A     66.6              0.817701  3.5       
rs746710   G/C     51.5              0.614368  0.0       
rs1430090  T/G     70.0              0.025180  1.6       
rs6737251  C/T     69.3              0.235996  0.3       
rs11685217 C/T     80.1              0.009462  4.5       
rs1430097  C/A     65.1              0.738166  1.0       
rs10496465 A/G     85.8              0.917997  0.6       
rs3756688  T/C     63.9              0.154632  0.6       
rs2303063  A/G     53.0              0.722069  1.1       
rs1422993  G/T     75.3              0.250093  0.0       
rs2400478  G/A     62.6              0.256786  0.9       
rs714588   A/G     54.9              0.838329  0.8       
rs1023555  T/A     76.8              0.943443  0.5       
rs898070   G/A     62.6              1.000000  0.6       
rs963218   C/T     53.2              0.389387  0.3       
rs1419835  C/T     78.2              0.505391  0.6       
rs765023   T/C     64.5              0.030513  6.9       
rs1345267  A/G     61.0              0.112183  0.1       
rs324381   G/A     64.6              0.242223 11.6       
rs184448   T/G     55.9              0.008446  2.2       
rs324396   C/T     71.2              0.197291  0.3       
rs324957   G/A     57.0              0.007417  0.4       
rs324960   C/T     66.6              0.077777  1.1       
rs10486657 C/T     81.3              0.672703  4.3       
rs324981   A/T     53.2              0.048438  0.2       
rs1419780  C/G     80.8              0.569652  0.2       
rs325462   T/A     51.0              0.337862  0.3       
rs727162   G/C     78.5              0.765708  0.0       
rs10250709 G/A     65.4              0.266434  0.0       
rs6958905  T/C     64.7              0.377472  0.4       
rs10238983 T/C     75.7              0.216435  0.4       
rs4941643  A/G     54.1              0.635887  7.2       
rs3794381  C/G     71.7              0.652089  7.1       
rs2031532  G/A     65.0              0.911918  0.0       
rs2247119  T/C     71.5              0.457710  0.5       
rs8000149  T/C     63.2              0.588077  0.4       
rs2274276  G/C     57.0              0.571386  0.6       
rs7332573  G/T     91.5              0.869947  1.5       
rs3829366  T/A     51.6              0.722626  1.3       
rs6084432  G/A     83.7              0.266716  0.6       
rs512625   G/A     69.5              0.905395  0.4       
rs3918395  G/T     86.8              0.508732  1.2       
rs2787095  G/C     60.2              0.102053  0.8       
rs2853215  G/A     73.0              0.249516  0.2       </code></pre>
<p>showing the SNP labels with minor/major allele format, the major allele frequency the HWE test and the percentage of missing genotypes. Missing values can be further explored plotting with</p>
<div class="sourceCode" id="cb164"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb164-1"><a href="single-association-analysis.html#cb164-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotMissing</span>(asthma.s, <span class="at">print.labels.SNPs =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plotMissing"></span>
<img src="fig/plotMissing-1.png" alt="Missing genotypes. Black squares shows missing genotuype information of asthma data example." width="672" />
<p class="caption">
Figure 3.3: Missing genotypes. Black squares shows missing genotuype information of asthma data example.
</p>
</div>
<p>This plot can be used to inspect if missing values appear randomly across individuals and SNPs. In our case, we can see that the missing pattern may be considered random, except for three clusters in consecutive SNPs (large black squares). These individuals should be further checked for possible problems with genotyping.</p>
</div>
<div id="hardy-weinberg-equilibrium" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Hardy-Weinberg equilibrium</h2>
<p>Genotyping of SNPs needs to pass quality control measures. Aside from technical details that need to be considered for filtering SNPs with low quality, genotype calling error can be detected by a HWE test. The test compares the observed genotype frequencies with those expected under random mating, which follows when the SNPs are in the absence of selection, mutation, genetic drift, or other forces. Therefore, HWE must be checked only in controls. There are several tests described in the literature to verify HWE. In <code>SNPassoc</code> HWE is tested for all the bi-allelic SNP markers using a fast exact test <span class="citation">(<a href="#ref-wigginton2005note" role="doc-biblioref">Wigginton, Cutler, and Abecasis 2005</a>)</span> implemented in the <em>tableHWE</em> function.</p>
<div class="sourceCode" id="cb165"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb165-1"><a href="single-association-analysis.html#cb165-1" aria-hidden="true" tabindex="-1"></a>hwe <span class="ot">&lt;-</span> <span class="fu">tableHWE</span>(asthma.s)</span>
<span id="cb165-2"><a href="single-association-analysis.html#cb165-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(hwe)</span></code></pre></div>
<pre><code>           HWE (p value)
rs4490198      0.1741325
rs4849332      0.5220596
rs1367179      0.7381531
rs11123242     0.9328981
rs13014858     0.3511162
rs1430094      0.3055089</code></pre>
<p>We observe that the first SNPs in the dataset are under HWE since their P-values rejecting the HWE hypothesis (null hypothesis) are larger than 0.05. However, when tested in control samples only, by stratifying by cases and controls</p>
<div class="sourceCode" id="cb167"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb167-1"><a href="single-association-analysis.html#cb167-1" aria-hidden="true" tabindex="-1"></a>hwe2 <span class="ot">&lt;-</span> <span class="fu">tableHWE</span>(asthma.s, casecontrol)</span>
<span id="cb167-2"><a href="single-association-analysis.html#cb167-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb167-3"><a href="single-association-analysis.html#cb167-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SNPs is HWE in the whole sample but not controls</span></span>
<span id="cb167-4"><a href="single-association-analysis.html#cb167-4" aria-hidden="true" tabindex="-1"></a>snpNHWE <span class="ot">&lt;-</span> hwe2[,<span class="dv">1</span>]<span class="sc">&gt;</span><span class="fl">0.05</span> <span class="sc">&amp;</span> hwe2[,<span class="dv">2</span>]<span class="sc">&lt;</span><span class="fl">0.05</span></span>
<span id="cb167-5"><a href="single-association-analysis.html#cb167-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(hwe2)[snpNHWE]</span></code></pre></div>
<pre><code>[1] &quot;rs1345267&quot;</code></pre>
<div class="sourceCode" id="cb169"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb169-1"><a href="single-association-analysis.html#cb169-1" aria-hidden="true" tabindex="-1"></a>hwe2[snpNHWE,]</span></code></pre></div>
<pre><code>all groups          0          1 
0.11218285 0.04956349 0.81604706 </code></pre>
<p>We see that rs1345267 is not in HWE within controls because its P-value is &lt;0.05. Notice that one is interested in keeping those SNPsthat do not reject the null hypothesis. As several SNPs are tested, multiple comparisons must be considered. In this particular setting, a threshold of 0.001 is normally considered. As a quality control measure, it is not necessary to be as conservative as in those situations where false discovery rates need to be controlled.</p>
<p>SNPs that do not pass the HWE test must be removed form further analyses. We can recall <em>setupSNP</em> and indicate the columns of the SNPs to be kept</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="single-association-analysis.html#cb171-1" aria-hidden="true" tabindex="-1"></a>snps.ok <span class="ot">&lt;-</span> <span class="fu">rownames</span>(hwe2)[hwe2[,<span class="dv">2</span>]<span class="sc">&gt;=</span><span class="fl">0.001</span>]</span>
<span id="cb171-2"><a href="single-association-analysis.html#cb171-2" aria-hidden="true" tabindex="-1"></a>pos <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">colnames</span>(asthma)<span class="sc">%in%</span>snps.ok, <span class="at">useNames =</span> <span class="cn">FALSE</span>)</span>
<span id="cb171-3"><a href="single-association-analysis.html#cb171-3" aria-hidden="true" tabindex="-1"></a>asthma.s <span class="ot">&lt;-</span> <span class="fu">setupSNP</span>(asthma, pos, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span></code></pre></div>
<p>Note that in the variable <em>pos</em> we redefine the SNP variables to be considered as class <em>snp</em>.</p>
</div>
<div id="single-snp-association-analysis" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Single SNP association analysis</h2>
<p>We are interested in finding those SNPs associated with asthma status that is encoded in the variable <em>casecontrol</em>. We first illustrate the association between case-control status and the SNP rs1422993. The association analysis for all genetic models is performed by the function <em>association</em> that regresses <em>casecontrol</em> on the variable rs1422993 from the dataset <em>asthma.s</em> that already contains the SNP variables of class <em>snp</em>.</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="single-association-analysis.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="fu">association</span>(casecontrol <span class="sc">~</span> rs1422993, <span class="at">data =</span> asthma.s)</span></code></pre></div>
<pre><code>
SNP: rs1422993  adjusted by: 
                0    %   1    %   OR lower upper  p-value  AIC
Codominant                                                    
G/G           730 59.0 173 50.9 1.00             0.017768 1642
G/T           425 34.3 145 42.6 1.44  1.12  1.85              
T/T            83  6.7  22  6.5 1.12  0.68  1.84              
Dominant                                                      
G/G           730 59.0 173 50.9 1.00             0.007826 1642
G/T-T/T       508 41.0 167 49.1 1.39  1.09  1.77              
Recessive                                                     
G/G-G/T      1155 93.3 318 93.5 1.00             0.877863 1649
T/T            83  6.7  22  6.5 0.96  0.59  1.57              
Overdominant                                                  
G/G-T/T       813 65.7 195 57.4 1.00             0.005026 1641
G/T           425 34.3 145 42.6 1.42  1.11  1.82              
log-Additive                                                  
0,1,2        1238 78.5 340 21.5 1.22  1.01  1.47 0.040151 1644</code></pre>
<p>The function <em>association</em> follows the usual syntax of R modelling functions with the difference that the variables in the model that are of class <em>snp</em> are tested using different genetic models. In our example, we observe that all genetic models but the recessive one are statistically significant. <em>association</em> also fits the overdominant model, which compares the two homozygous genotypes versus the heterozygous one. This genetic model of inheritance is biologically rare although it has been linked to sickle cell anemia in humans. The result table describes the number of individuals in each genotype across cases and controls. The ORs and CI-95% are also computed. The last column describes the AIC (Akaike information criteria) that can be used to decide which is the best model of inheritance; the lower the better the model is.</p>
<p>In the example, one may conclude that rs1422993 is associated with asthma and that, for instance, the risk of being asthmatic is 39% higher in people having at least one alternative allele (T) with respect to individuals having none (dominant model). This risk is statistically significant since the CI-95% does not contain 1, the P-value is 0.0078&lt;0.022, or the P-value of the max-statistics is 0.01</p>
<div class="sourceCode" id="cb174"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb174-1"><a href="single-association-analysis.html#cb174-1" aria-hidden="true" tabindex="-1"></a><span class="fu">maxstat</span>(asthma.s<span class="sc">$</span>casecontrol, asthma.s<span class="sc">$</span>rs1422993)</span></code></pre></div>
<pre><code>     dominant recessive log-additive MAX-statistic Pr(&gt;z)  
[1,]    7.073     0.024        4.291         7.073 0.0166 *
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>If an expected model of inheritance is hypothesized, the association analysis for the model can be specified in the argument <em>model</em>, which by default test all models,</p>
<div class="sourceCode" id="cb176"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb176-1"><a href="single-association-analysis.html#cb176-1" aria-hidden="true" tabindex="-1"></a><span class="fu">association</span>(casecontrol <span class="sc">~</span> rs1422993, asthma.s, <span class="at">model=</span><span class="st">&quot;dominant&quot;</span>)</span></code></pre></div>
<pre><code>
SNP: rs1422993  adjusted by: 
           0  %   1    %   OR lower upper  p-value  AIC
Dominant                                               
G/G      730 59 173 50.9 1.00             0.007826 1642
G/T-T/T  508 41 167 49.1 1.39  1.09  1.77              </code></pre>
<p>Association tests are typically adjusted by covariates, which are incorporated in the model in the usual form</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="single-association-analysis.html#cb178-1" aria-hidden="true" tabindex="-1"></a><span class="fu">association</span>(casecontrol <span class="sc">~</span> rs1422993 <span class="sc">+</span> country <span class="sc">+</span> smoke, asthma.s)</span></code></pre></div>
<pre><code>
SNP: rs1422993  adjusted by: country smoke 
                0    %   1    %   OR lower upper p-value  AIC
Codominant                                                   
G/G           728 59.1 173 51.0 1.00             0.06957 1407
G/T           423 34.3 144 42.5 1.38  1.05  1.82             
T/T            81  6.6  22  6.5 1.07  0.62  1.85             
Dominant                                                     
G/G           728 59.1 173 51.0 1.00             0.03380 1406
G/T-T/T       504 40.9 166 49.0 1.33  1.02  1.73             
Recessive                                                    
G/G-G/T      1151 93.4 317 93.5 1.00             0.80821 1411
T/T            81  6.6  22  6.5 0.94  0.55  1.60             
Overdominant                                                 
G/G-T/T       809 65.7 195 57.5 1.00             0.02163 1406
G/T           423 34.3 144 42.5 1.37  1.05  1.79             
log-Additive                                                 
0,1,2        1232 78.4 339 21.6 1.19  0.96  1.46 0.10926 1408</code></pre>
<p>ORs for stratified analysis on given categorical covariates are used to verify whether the risk is constant across groups</p>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="single-association-analysis.html#cb180-1" aria-hidden="true" tabindex="-1"></a><span class="fu">association</span>(casecontrol <span class="sc">~</span> rs1422993 <span class="sc">+</span> survival<span class="sc">::</span><span class="fu">strata</span>(gender), asthma.s)</span></code></pre></div>
<pre><code>
SNP: rs1422993  adjusted by: survival::strata(gender) 
                0    %   1    %   OR lower upper  p-value  AIC
Codominant                                                    
G/G           730 59.0 173 50.9 1.00             0.022940 1634
G/T           425 34.3 145 42.6 1.42  1.11  1.83              
T/T            83  6.7  22  6.5 1.09  0.66  1.80              
Dominant                                                      
G/G           730 59.0 173 50.9 1.00             0.011144 1633
G/T-T/T       508 41.0 167 49.1 1.37  1.07  1.74              
Recessive                                                     
G/G-G/T      1155 93.3 318 93.5 1.00             0.805330 1640
T/T            83  6.7  22  6.5 0.94  0.58  1.53              
Overdominant                                                  
G/G-T/T       813 65.7 195 57.4 1.00             0.006378 1632
G/T           425 34.3 145 42.6 1.41  1.10  1.80              
log-Additive                                                  
0,1,2        1238 78.5 340 21.5 1.21  1.00  1.46 0.055231 1636</code></pre>
<p>We can see, for instance, that the dominant model is significant only in males. The subset argument allows fitting the model in a subgroup of individuals</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="single-association-analysis.html#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="fu">association</span>(casecontrol <span class="sc">~</span> rs1422993, asthma.s, </span>
<span id="cb182-2"><a href="single-association-analysis.html#cb182-2" aria-hidden="true" tabindex="-1"></a>                <span class="at">subset=</span>country<span class="sc">==</span><span class="st">&quot;Spain&quot;</span>)</span></code></pre></div>
<pre><code>
SNP: rs1422993  adjusted by: 
               0    %  1    %   OR lower upper p-value   AIC
Codominant                                                  
G/G          179 54.6 22 44.9 1.00              0.3550 295.2
G/T          125 38.1 24 49.0 1.56  0.84  2.91              
T/T           24  7.3  3  6.1 1.02  0.28  3.66              
Dominant                                                    
G/G          179 54.6 22 44.9 1.00              0.2059 293.7
G/T-T/T      149 45.4 27 55.1 1.47  0.81  2.70              
Recessive                                                   
G/G-G/T      304 92.7 46 93.9 1.00              0.7576 295.2
T/T           24  7.3  3  6.1 0.83  0.24  2.85              
Overdominant                                                
G/G-T/T      203 61.9 25 51.0 1.00              0.1502 293.2
G/T          125 38.1 24 49.0 1.56  0.85  2.85              
log-Additive                                                
0,1,2        328 87.0 49 13.0 1.23  0.77  1.96  0.3816 294.5</code></pre>
<p>These analyses can be also be performed in quantitative traits, such as body mass index, since <em>association</em> function automatically selects the error distribution of the regression analysis (either Gaussian or binomial).</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="single-association-analysis.html#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="fu">association</span>(bmi <span class="sc">~</span> rs1422993, asthma.s) </span></code></pre></div>
<pre><code>
SNP: rs1422993  adjusted by: 
                n    me     se       dif   lower  upper p-value  AIC
Codominant                                                          
G/G           896 25.53 0.1446  0.000000                 0.9069 9069
G/T           565 25.50 0.1834 -0.027059 -0.4874 0.4332             
T/T           105 25.71 0.4676  0.178076 -0.7057 1.0619             
Dominant                                                            
G/G           896 25.53 0.1446  0.000000                 0.9818 9067
G/T-T/T       670 25.54 0.1710  0.005089 -0.4324 0.4426             
Recessive                                                           
G/G-G/T      1461 25.52 0.1135  0.000000                 0.6694 9067
T/T           105 25.71 0.4676  0.188540 -0.6769 1.0540             
Overdominant                                                        
G/G-T/T      1001 25.55 0.1383  0.000000                 0.8424 9067
G/T           565 25.50 0.1834 -0.045739 -0.4965 0.4050             
log-Additive                                                        
0,1,2                           0.033951 -0.3153 0.3832  0.8489 9067</code></pre>
<p>For BMI, <em>association</em> tests whether the difference between means is statistically significant, rather than computing an OR.</p>
<p>For multiple SNP data, our objective is to identify the variants that are significantly associated with the trait. The most basic strategy is, therefore, to fit an association test like the one described above for each of the SNPs in the dataset and determine which of those associations are significant. The massive univariate testing is the most widely used analysis method for <em>omic</em> data because of its simplicity. In <em>SNPassoc</em>, this type of analysis is done with the function <em>WGassociation</em></p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="single-association-analysis.html#cb186-1" aria-hidden="true" tabindex="-1"></a>ans <span class="ot">&lt;-</span> <span class="fu">WGassociation</span>(casecontrol, <span class="at">data=</span>asthma.s)</span>
<span id="cb186-2"><a href="single-association-analysis.html#cb186-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ans)</span></code></pre></div>
<pre><code>           comments codominant dominant recessive overdominant
rs4490198         -    0.52765  0.29503   0.96400      0.29998
rs4849332         -    0.96912  0.92986   0.84806      0.82327
rs1367179         -    0.62775  0.59205   0.35786      0.86419
rs11123242        -    0.68622  0.67596   0.39801      0.92878
rs13014858        -    0.52578  0.26739   0.88011      0.34966
rs1430094         -    0.13375  0.10569   0.54432      0.04490
           log-additive
rs4490198       0.49506
rs4849332       0.97049
rs1367179       0.43994
rs11123242      0.52009
rs13014858      0.40897
rs1430094       0.36611</code></pre>
<p>Here, only the outcome is required in the formula argument (first argument) since the function successively calls <em>association</em> on each of the variables of class <em>snp</em> within <em>data</em>. The function returns the P-values of association of each SNP under each genetic model. Covariates can also be introduced in the model</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="single-association-analysis.html#cb188-1" aria-hidden="true" tabindex="-1"></a>ans.adj <span class="ot">&lt;-</span> <span class="fu">WGassociation</span>(casecontrol <span class="sc">~</span> country <span class="sc">+</span> smoke, asthma.s)</span>
<span id="cb188-2"><a href="single-association-analysis.html#cb188-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ans.adj)</span></code></pre></div>
<p><code>SNPassoc</code> is computationally limited on large genomic data. The computing time can be reduced by parallelization, specifying in the argument <em>mc.cores</em> the number of computing cores to be used. Alternatively, the function <em>scanWGassociation</em>, a C compiled function, can be used to compute a predetermined genetic model across all SNPs, passed in the argument <em>model</em>, which by default is the additive model</p>
<pre class="more"><code>ans.fast &lt;- scanWGassociation(casecontrol, asthma.s)</code></pre>
<p><strong>NOTE</strong>: This function is not available on the <code>SNPassoc</code> version available on CRAN. The user can install the development version available on GitHub to get access to this function just executing</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="single-association-analysis.html#cb190-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;isglobal-brge/SNPassoc&quot;</span>)</span></code></pre></div>
<p>The P-values obtained from massive univariate analyses are visualized with the generic <em>plot</em> function</p>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="single-association-analysis.html#cb191-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ans)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plotGWAS"></span>
<img src="fig/plotGWAS-1.png" alt="Manhattan-type plots for different genetic models. P-values in -log10 scale to assess the association between case-control status and SNPs in the asthma example." width="672" />
<p class="caption">
Figure 3.4: Manhattan-type plots for different genetic models. P-values in -log10 scale to assess the association between case-control status and SNPs in the asthma example.
</p>
</div>
<p>This produces a Manhattan plot of the -log10(P-values) for all the SNPs over all models. It shows the nominal level of significance and the Bonferroni level, which is the level corrected by the multiple testing across all SNPs. The overall hypothesis of massive univariate association tests is whether there is any SNP that is significantly associated with the phenotype. As multiple SNPs are tested, the probability of finding at least one significant finding increases if we do not lower the significance threshold. The Bonferroni correction lowers the threshold by the number of SNPs tested (0.0001=0.05/51). In the Manhattan plotof our analysis, we see that no SNP is significant at the Bonferroni level, and therefore there is no SNP that is significantly associated with asthma.</p>
<p>Maximum-statistic <span class="citation">(see <a href="#ref-gonzalez2008maximizing" role="doc-biblioref">González et al. 2008</a>)</span> can also be used to test association between asthma status and SNPs</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="single-association-analysis.html#cb192-1" aria-hidden="true" tabindex="-1"></a>ans.max <span class="ot">&lt;-</span> <span class="fu">maxstat</span>(asthma.s, casecontrol)</span>
<span id="cb192-2"><a href="single-association-analysis.html#cb192-2" aria-hidden="true" tabindex="-1"></a>ans.max</span></code></pre></div>
<pre><code>           dominant recessive log-additive MAX-statistic  Pr(&gt;z)   
rs4490198     1.097     0.002        0.466         1.097 0.50189   
rs4849332     0.008     0.037        0.001         0.037 0.97609   
rs1367179     0.287     0.845        0.602         0.845 0.58888   
rs11123242    0.175     0.714        0.417         0.714 0.63964   
rs13014858    1.230     0.023        0.683         1.230 0.46295   
rs1430094     2.617     0.368        0.821         2.617 0.20673   
rs1430093     1.051     0.042        0.743         1.051 0.51864   
rs746710      0.728     0.679        1.051         1.051 0.51628   
rs1430090     0.172     0.463        0.000         0.463 0.74308   
rs6737251     0.143     0.156        0.217         0.217 0.86919   
rs11685217    0.894     0.030        0.705         0.894 0.56883   
rs1430097     0.003     0.183        0.029         0.183 0.88853   
rs10496465    0.003     0.020        0.008         0.020 0.98739   
rs3756688     0.016     0.738        0.266         0.738 0.62640   
rs2303063     0.060     1.271        0.658         1.271 0.45273   
rs1422993     7.073     0.024        4.291         7.073 0.01821 * 
rs2400478     1.662     0.056        1.055         1.662 0.35752   
rs714588      0.659     0.061        0.150         0.659 0.65815   
rs1023555     0.221     0.104        0.261         0.261 0.84692   
rs898070      0.020     1.794        0.346         1.794 0.33162   
rs963218      0.165     0.190        0.263         0.263 0.84295   
rs1419835     1.084     0.775        0.295         1.084 0.51076   
rs765023      1.959     0.526        0.483         1.959 0.30142   
rs1345267     2.582     0.003        1.383         2.582 0.21106   
rs324381      0.175     0.402        0.377         0.402 0.77466   
rs184448      9.710     2.026        8.253         9.710 0.00339 **
rs324396      2.312     0.188        1.051         2.312 0.24921   
rs324957      7.598     2.134        7.157         7.598 0.01437 * 
rs324960      2.901     7.928        6.443         7.928 0.01159 * 
rs10486657    0.054     0.500        0.001         0.500 0.72957   
rs324981      3.168     2.431        4.270         4.270 0.08105 . 
rs1419780     0.123     1.193        0.005         1.193 0.47806   
rs325462      1.545     1.645        2.412         2.412 0.23342   
rs727162      3.022     0.643        3.074         3.074 0.16080   
rs10250709    0.737     0.007        0.360         0.737 0.62798   
rs6958905     0.490     0.099        0.447         0.490 0.73268   
rs10238983    0.094     0.692        0.328         0.692 0.64821   
rs4941643     1.294     0.006        0.465         1.294 0.44652   
rs3794381     0.127     0.998        0.489         0.998 0.53565   
rs2031532     0.025     0.240        0.127         0.240 0.85704   
rs2247119     0.379     0.000        0.229         0.379 0.78427   
rs8000149     0.020     0.043        0.043         0.043 0.97280   
rs2274276     0.005     0.034        0.023         0.034 0.97788   
rs7332573     1.247     1.703        1.825         1.825 0.32966   
rs3829366     1.137     0.448        0.069         1.137 0.49124   
rs6084432     2.999     0.848        3.279         3.279 0.14197   
rs512625      2.180     0.030        1.119         2.180 0.26476   
rs3918395     0.903     0.414        0.455         0.903 0.57149   
rs2787095     0.178     0.069        0.184         0.184 0.88729   
rs2853215     0.686     0.931        1.130         1.130 0.49261   
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1</code></pre>
<p>We note that even under the max-statistics none of the SNPs tested is significant under the Bonferroni correction (&lt;0.0001) for multiple SNP testing</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="single-association-analysis.html#cb194-1" aria-hidden="true" tabindex="-1"></a><span class="co">#minimum P-value across SNPs</span></span>
<span id="cb194-2"><a href="single-association-analysis.html#cb194-2" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(ans.max[<span class="st">&quot;Pr(&gt;z)&quot;</span>,])</span></code></pre></div>
<pre><code>[1] 0.003392342</code></pre>
<p>Information for specific association models for given SNPs can also be retrieved with <em>WGstats</em></p>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="single-association-analysis.html#cb196-1" aria-hidden="true" tabindex="-1"></a>infoTable <span class="ot">&lt;-</span> <span class="fu">WGstats</span>(ans)</span></code></pre></div>
<p>Therefore, we can have access to the results for a given SNP by</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="single-association-analysis.html#cb197-1" aria-hidden="true" tabindex="-1"></a>infoTable<span class="sc">$</span>rs1422993</span></code></pre></div>
<pre><code>
SNP: rs1422993  adjusted by: 
                0    %   1    %   OR lower upper  p-value  AIC
Codominant                                                    
G/G           730 59.0 173 50.9 1.00             0.017768 1642
G/T           425 34.3 145 42.6 1.44  1.12  1.85              
T/T            83  6.7  22  6.5 1.12  0.68  1.84              
Dominant                                                      
G/G           730 59.0 173 50.9 1.00             0.007826 1642
G/T-T/T       508 41.0 167 49.1 1.39  1.09  1.77              
Recessive                                                     
G/G-G/T      1155 93.3 318 93.5 1.00             0.877863 1649
T/T            83  6.7  22  6.5 0.96  0.59  1.57              
Overdominant                                                  
G/G-T/T       813 65.7 195 57.4 1.00             0.005026 1641
G/T           425 34.3 145 42.6 1.42  1.11  1.82              
log-Additive                                                  
0,1,2        1238 78.5 340 21.5 1.22  1.01  1.47 0.040151 1644</code></pre>
<p>recovering our previous results given by <em>association</em> function.</p>
<p><strong>NOTE</strong>: The R output of specific association analyses can be exported into LaTeX by using <em>getNiceTable</em> function and <em>xtable</em> R package. The following code creates a table for the SNPs rs1422993 and rs184448</p>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="single-association-analysis.html#cb199-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(xtable)</span>
<span id="cb199-2"><a href="single-association-analysis.html#cb199-2" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">getNiceTable</span>(ans[<span class="fu">c</span>(<span class="st">&quot;rs1422993&quot;</span>, <span class="st">&quot;rs184448&quot;</span>)])</span>
<span id="cb199-3"><a href="single-association-analysis.html#cb199-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb199-4"><a href="single-association-analysis.html#cb199-4" aria-hidden="true" tabindex="-1"></a>nlines <span class="ot">&lt;-</span> <span class="fu">attr</span>(out, <span class="st">&quot;nlines&quot;</span>)</span>
<span id="cb199-5"><a href="single-association-analysis.html#cb199-5" aria-hidden="true" tabindex="-1"></a>hlines <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="sc">-</span><span class="dv">1</span>, <span class="dv">0</span>, <span class="fu">cumsum</span>(nlines<span class="sc">+</span><span class="dv">1</span>), <span class="fu">nrow</span>(out), <span class="fu">nrow</span>(out))</span>
<span id="cb199-6"><a href="single-association-analysis.html#cb199-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb199-7"><a href="single-association-analysis.html#cb199-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">xtable</span>(out, <span class="at">caption=</span><span class="st">&#39;Genetic association using</span></span>
<span id="cb199-8"><a href="single-association-analysis.html#cb199-8" aria-hidden="true" tabindex="-1"></a><span class="st">                different genetic models from asthma </span></span>
<span id="cb199-9"><a href="single-association-analysis.html#cb199-9" aria-hidden="true" tabindex="-1"></a><span class="st">                data example of rs1422993 and rs184448 </span></span>
<span id="cb199-10"><a href="single-association-analysis.html#cb199-10" aria-hidden="true" tabindex="-1"></a><span class="st">                SNPs obtained with SNPassoc.&#39;</span>,</span>
<span id="cb199-11"><a href="single-association-analysis.html#cb199-11" aria-hidden="true" tabindex="-1"></a>             <span class="at">label =</span> <span class="st">&#39;tab-2SNPs&#39;</span>),</span>
<span id="cb199-12"><a href="single-association-analysis.html#cb199-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">tabular.enviroment=</span><span class="st">&quot;longtable&quot;</span>, <span class="at">file=</span><span class="st">&quot;tableSNPs&quot;</span>,</span>
<span id="cb199-13"><a href="single-association-analysis.html#cb199-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">floating=</span><span class="cn">FALSE</span>,  <span class="at">include.rownames =</span> <span class="cn">FALSE</span>, </span>
<span id="cb199-14"><a href="single-association-analysis.html#cb199-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">hline.after=</span> hlines, <span class="at">sanitize.text.function=</span>identity)</span></code></pre></div>
</div>
<div id="gene-environment-and-gene-gene-interactions" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Gene-environment and gene-gene interactions</h2>
<p>Gene-enviroment (GxE) analyses can be performed within <code>SNPassoc</code> using <em>association</em> function. Assume that we are interested in testing whether the risk of rs1422993 for asthma under the dominant model is different among smokers (variable smoke; 0=never, 1=ever). This code fits a model with an interaction term where the environmental variable is required to be a factor factor variable.</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="single-association-analysis.html#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="fu">association</span>(casecontrol <span class="sc">~</span> <span class="fu">dominant</span>(rs1422993)<span class="sc">*</span><span class="fu">factor</span>(smoke), </span>
<span id="cb200-2"><a href="single-association-analysis.html#cb200-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">data=</span>asthma.s)</span></code></pre></div>
<pre><code>
      SNP: dominant(rs1422993  adjusted by: 
 Interaction 
---------------------
          0      OR lower upper   1      OR lower upper   0  1    2
G/G     273 86 1.00    NA    NA 213 44 0.66  0.44  0.98 182 34 0.59
G/T-T/T 210 77 1.16  0.82  1.66 144 51 1.12  0.75  1.68 115 24 0.66
        lower upper  0 1    3 lower upper  0 1    4 lower upper
G/G      0.38  0.92 40 2 0.16  0.04  0.67 20 7 1.11  0.45  2.72
G/T-T/T  0.40  1.09 19 7 1.17  0.48  2.88 16 7 1.39  0.55  3.49

p interaction: 0.13712 

 factor(smoke) within dominant(rs1422993 
---------------------
G/G 
    0  1   OR lower upper
0 273 86 1.00    NA    NA
1 213 44 0.66  0.44  0.98
2 182 34 0.59  0.38  0.92
3  40  2 0.16  0.04  0.67
4  20  7 1.11  0.45  2.72

G/T-T/T 
    0  1   OR lower upper
0 210 77 1.00    NA    NA
1 144 51 0.97  0.64  1.46
2 115 24 0.57  0.34  0.95
3  19  7 1.00  0.41  2.48
4  16  7 1.19  0.47  3.01

p trend: 0.13712 

 dominant(rs1422993 within factor(smoke) 
---------------------
0 
          0  1   OR lower upper
G/G     273 86 1.00    NA    NA
G/T-T/T 210 77 1.16  0.82  1.66

1 
          0  1   OR lower upper
G/G     213 44 1.00    NA    NA
G/T-T/T 144 51 1.71  1.09   2.7

2 
          0  1   OR lower upper
G/G     182 34 1.00    NA    NA
G/T-T/T 115 24 1.12  0.63  1.98

3 
         0 1   OR lower upper
G/G     40 2 1.00    NA    NA
G/T-T/T 19 7 7.37   1.4 38.89

4 
         0 1   OR lower upper
G/G     20 7 1.00    NA    NA
G/T-T/T 16 7 1.25  0.36  4.31

p trend: 0.278 </code></pre>
<p>The result is an interaction table showing that the risk of individuals carrying the T allele increases the risk of asthma in never smokers (OR=1.35; CI: 1.02-1.79) while it is not significant in ever smokers (OR=0.95; CI: 0.63-1.42). However, the interaction is not statistically significant (<span class="math inline">\(P\)</span>-interaction=0.8513). The output also shows the stratified ORs that can help in interpreting the results.</p>
<p>In a similar way, gene-gene interaction (GxG) of a given SNP epistasis model can also be fitted using the same function. In that case, the genetic model of the interacting SNP must be indicated in the <em>model.inteaction</em> argument.</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="single-association-analysis.html#cb202-1" aria-hidden="true" tabindex="-1"></a><span class="fu">association</span>(casecontrol <span class="sc">~</span> rs1422993<span class="sc">*</span><span class="fu">factor</span>(rs184448), </span>
<span id="cb202-2"><a href="single-association-analysis.html#cb202-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">data=</span>asthma.s, <span class="at">model.interaction =</span> <span class="st">&quot;dominant&quot;</span> )</span></code></pre></div>
<pre><code>
      SNP: rs1422993  adjusted by: 
 Interaction 
---------------------
        T/T      OR lower upper T/G      OR lower upper   0  1  G/G
G/G     227 43 1.00    NA    NA 359 96 1.41  0.95  2.10 128 30 1.24
G/T-T/T 154 33 1.13  0.69  1.86 265 93 1.85  1.24  2.77  78 38 2.57
        lower upper
G/G      0.74  2.07
G/T-T/T  1.55  4.27

p interaction: 0.24499 

 factor(rs184448) within rs1422993 
---------------------
G/G 
      0  1   OR lower upper
T/T 227 43 1.00    NA    NA
T/G 359 96 1.41  0.95  2.10
G/G 128 30 1.24  0.74  2.07

G/T-T/T 
      0  1   OR lower upper
T/T 154 33 1.00    NA    NA
T/G 265 93 1.64  1.05  2.55
G/G  78 38 2.27  1.32  3.90

p trend: 0.24499 

 rs1422993 within factor(rs184448) 
---------------------
T/T 
          0  1   OR lower upper
G/G     227 43 1.00    NA    NA
G/T-T/T 154 33 1.13  0.69  1.86

T/G 
          0  1   OR lower upper
G/G     359 96 1.00    NA    NA
G/T-T/T 265 93 1.31  0.95  1.82

G/G 
          0  1   OR lower upper
G/G     128 30 1.00    NA    NA
G/T-T/T  78 38 2.08  1.19  3.62

p trend: 0.12743 </code></pre>
<p>We observe that the interaction between these two SNPs is not statistically significant (P-value=0.24). However, the OR of GG genotype of rs184448 differs across individuals between the GG and GT-TT genotypes of rs1422993 (see ORs for GG in the second table of the output).</p>
<p>The user also can perform GxG for a set of SNPs using this code. Let us assume we are interested in assessing interaction between the SNPs that are significant at 10% level</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="single-association-analysis.html#cb204-1" aria-hidden="true" tabindex="-1"></a>ans <span class="ot">&lt;-</span> <span class="fu">WGassociation</span>(casecontrol, <span class="at">data=</span>asthma.s)</span>
<span id="cb204-2"><a href="single-association-analysis.html#cb204-2" aria-hidden="true" tabindex="-1"></a>mask <span class="ot">&lt;-</span> <span class="fu">apply</span>(ans, <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="fu">min</span>(x, <span class="at">na.rm=</span><span class="cn">TRUE</span>)<span class="sc">&lt;</span><span class="fl">0.1</span>)</span>
<span id="cb204-3"><a href="single-association-analysis.html#cb204-3" aria-hidden="true" tabindex="-1"></a>sig.snps <span class="ot">&lt;-</span> <span class="fu">names</span>(mask[mask])</span>
<span id="cb204-4"><a href="single-association-analysis.html#cb204-4" aria-hidden="true" tabindex="-1"></a>sig.snps</span></code></pre></div>
<pre><code> [1] &quot;rs1430094&quot; &quot;rs1422993&quot; &quot;rs765023&quot;  &quot;rs184448&quot;  &quot;rs324396&quot; 
 [6] &quot;rs324957&quot;  &quot;rs324960&quot;  &quot;rs324981&quot;  &quot;rs727162&quot;  &quot;rs6084432&quot;</code></pre>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="single-association-analysis.html#cb206-1" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">colnames</span>(asthma)<span class="sc">%in%</span>sig.snps)</span>
<span id="cb206-2"><a href="single-association-analysis.html#cb206-2" aria-hidden="true" tabindex="-1"></a>asthma.s2 <span class="ot">&lt;-</span> <span class="fu">setupSNP</span>(asthma, <span class="at">colSNPs =</span> idx, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb206-3"><a href="single-association-analysis.html#cb206-3" aria-hidden="true" tabindex="-1"></a>ans.int <span class="ot">&lt;-</span> <span class="fu">interactionPval</span>(casecontrol <span class="sc">~</span> <span class="dv">1</span>, <span class="at">data=</span>asthma.s2)</span>
<span id="cb206-4"><a href="single-association-analysis.html#cb206-4" aria-hidden="true" tabindex="-1"></a>ans.int</span></code></pre></div>
<pre><code>            rs1430094   rs1422993    rs765023    rs184448    rs324396
rs1430094 0.132526514 0.653457029 0.816154586 0.787386835 0.694311497
rs1422993 0.140433948 0.016719949 0.133131712 0.375246182 0.683959376
rs765023  0.171993821 0.144029395 0.182805988 0.961405419 0.194520134
rs184448  0.100074594 0.019963426 0.034666219 0.007969948 0.036825613
rs324396  0.204742661 0.190583402 0.131945703 0.344952544 0.209589962
rs324957  0.103680541 0.019820456 0.034446832 0.696623215 0.318903026
rs324960  0.107216765 0.024180405 0.174801032 0.291457716 0.584132142
rs324981  0.128489057 0.144077239 0.153521356 0.646139555 0.301782199
rs727162  0.258828613 0.240900610 0.188971595 0.173341993 0.120240909
rs6084432 0.214413938 0.187372912 0.187830183 0.264160389 0.252965105
             rs324957    rs324960    rs324981    rs727162   rs6084432
rs1430094 0.644661690 0.096913557 0.774083175 0.043758067 0.946641185
rs1422993 0.257661509 0.175946789 0.095251421 0.577504655 0.027437424
rs765023  0.437535350 0.614095008 0.117705086 0.217580773 0.896707843
rs184448  0.018290318 0.356124918 0.311031078 0.297592380 0.349332875
rs324396  0.232076552 0.747574191 0.064231657 0.361336095 0.388626207
rs324957  0.019578585 0.595796247 0.965290164 0.402781904 0.272919867
rs324960  0.404830873 0.017401694 0.168077057 0.406220478 0.725774374
rs324981  0.657378128 0.533095852 0.117043376 0.463436271 0.862291944
rs727162  0.174960014 0.500811687 0.167364803 0.206131411 0.885294485
rs6084432 0.174006927 0.146808311 0.210647392 0.203957720 0.193711884</code></pre>
<p>we can visualize the results by</p>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="single-association-analysis.html#cb208-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(ans.int)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plotInf"></span>
<img src="fig/plotInf-1.png" alt="Interaction plot. Interaction plot of SNPs significant al 10\% significant level (see help of 'interactionPval'  function to see what is represented in the plot). " width="672" />
<p class="caption">
Figure 3.5: Interaction plot. Interaction plot of SNPs significant al 10% significant level (see help of ‘interactionPval’ function to see what is represented in the plot).
</p>
</div>
</div>
<div id="haplotype-analysis" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> Haplotype analysis</h2>
<p>Genetic association studies can be extended from single SNP associations to haplotype associations. Several examples including different complex diseases can be found in . While alleles naturally occur in haplotypes, as they belong to chromosomes, the phase, or the knowledge of the chromosome an allele belongs to is lost in the genotyping process. As each genotype is measured with a different probe the phase between the alleles in a chromosome is broken. Consider for instance an individual for which one chromosome has alleles A and T at two different loci and alleles G or C at the second chromosome. The individual’s genotypes for the individual at the two loci are A/G and T/C, which are the same genotypes of another individual that has alleles A and C in one chromosome and G and T in the second chromosome. Clearly, if the pair A-T confers a risk to a disease, subject one is at risk while subject two is not, despite both of them having the same genotypes.</p>
<p>The only unequivocal method of resolving phase ambiguity is sequencing the chromosomes of individuals. However, given the correlational structure of SNPs, it is possible to estimate the probability of a particular haplotype in a subject. This can be done in a genetic association study where a number of cases and controls are genotyped. There are numerous methods to infer unobserved haplotypes, two of the most popular are maximum likelihood, implemented via the expectation-maximization (EM) algorithm , and a parsimony method . Recent methods based on Bayesian models have also been proposed . In addition, haplotypes inferences carry uncertainty, which should be considered in association analyses.</p>
<div id="haplotype-estimation" class="section level3" number="3.5.1">
<h3><span class="header-section-number">3.5.1</span> Haplotype estimation</h3>
<p>We now illustrate how to perform haplotype estimation from genotype data using the EM algorithm and how to integrate haplotype uncertainty when evaluating the association between traits and haplotypes. Haplotype inference is performed with <code>haplo.stats</code>, for which genotypes are encoded in a different format. <em>make.geno</em>, from <code>SNPassoc</code>, formats data for <code>haplo.stats</code>. The function <em>haplo.em</em> computes the haplotype frequency in the data for the SNPs of interest. Here we illustrate how to estimate haplotypes built from the SNPs rs714588, rs1023555 and rs898070:</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="single-association-analysis.html#cb209-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(haplo.stats)</span>
<span id="cb209-2"><a href="single-association-analysis.html#cb209-2" aria-hidden="true" tabindex="-1"></a>snpsH <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;rs714588&quot;</span>, <span class="st">&quot;rs1023555&quot;</span>,  <span class="st">&quot;rs898070&quot;</span>)</span>
<span id="cb209-3"><a href="single-association-analysis.html#cb209-3" aria-hidden="true" tabindex="-1"></a>genoH <span class="ot">&lt;-</span> <span class="fu">make.geno</span>(asthma.s, snpsH)</span>
<span id="cb209-4"><a href="single-association-analysis.html#cb209-4" aria-hidden="true" tabindex="-1"></a>em <span class="ot">&lt;-</span> <span class="fu">haplo.em</span>(genoH, <span class="at">locus.label =</span> snpsH, <span class="at">miss.val =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="cn">NA</span>))</span>
<span id="cb209-5"><a href="single-association-analysis.html#cb209-5" aria-hidden="true" tabindex="-1"></a>em</span></code></pre></div>
<pre><code>====================================================================== 
                              Haplotypes                               
====================================================================== 
  rs714588 rs1023555 rs898070 hap.freq
1        1         1        1  0.04090
2        1         1        2  0.02439
3        1         2        1  0.04265
4        1         2        2  0.44047
5        2         1        1  0.08271
6        2         1        2  0.08403
7        2         2        1  0.20794
8        2         2        2  0.07691
====================================================================== 
                               Details                                 
====================================================================== 
lnlike =  -4102.691 
lr stat for no LD =  774.5411 , df =  4 , p-val =  0 </code></pre>
<p>Coding the common and variant alleles as 1 and 2, we can see there are 8 possible haplotypes across the subjects and are listed with an estimated haplotype frequency. Clearly, the haplotypes are not equally probable, as expected from the high LD between the SNPs. In particular, we observe that haplotypes 4 and 7 are the most probable, accumulating 65% of the haplotype sample. <em>haplo.em</em> estimates for each subject the probability of a given haplotype in each of the subject’s chromosomes.</p>
</div>
<div id="haplotype-association" class="section level3" number="3.5.2">
<h3><span class="header-section-number">3.5.2</span> Haplotype association</h3>
<p>We then want to assess if any of these haplotypes significantly associates with asthma. The <em>haplo.glm</em> fits a regression model between the phenotype and the haplotypes, incorporating the uncertainty for the probable haplotypes of individuals. The function <em>intervals</em> of <code>SNPassoc</code> provides a nice summary of the results</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="single-association-analysis.html#cb211-1" aria-hidden="true" tabindex="-1"></a>trait <span class="ot">&lt;-</span> asthma.s<span class="sc">$</span>casecontrol</span>
<span id="cb211-2"><a href="single-association-analysis.html#cb211-2" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">haplo.glm</span>(trait <span class="sc">~</span> genoH,           </span>
<span id="cb211-3"><a href="single-association-analysis.html#cb211-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">family=</span><span class="st">&quot;binomial&quot;</span>, </span>
<span id="cb211-4"><a href="single-association-analysis.html#cb211-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">locus.label=</span>snpsH,</span>
<span id="cb211-5"><a href="single-association-analysis.html#cb211-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">allele.lev=</span><span class="fu">attributes</span>(genoH)<span class="sc">$</span>unique.alleles,</span>
<span id="cb211-6"><a href="single-association-analysis.html#cb211-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">control =</span> <span class="fu">haplo.glm.control</span>(<span class="at">haplo.freq.min=</span><span class="fl">0.05</span>))   </span>
<span id="cb211-7"><a href="single-association-analysis.html#cb211-7" aria-hidden="true" tabindex="-1"></a><span class="fu">intervals</span>(mod)</span></code></pre></div>
<pre><code>               freq   or   95%   C.I.   P-val 
         ATG 0.4405   1.00 Reference haplotype 
         GAA 0.0827   1.09 (   0.77 -   1.56 )  0.6160 
         GAG 0.0841   0.99 (   0.69 -   1.42 )  0.9642 
         GTA 0.2080   1.06 (   0.84 -   1.34 )  0.6379 
         GTG 0.0769   1.09 (   0.76 -   1.58 )  0.6367 
  genoH.rare 0.1079   1.15 (   0.83 -   1.59 )  0.3968 </code></pre>
<p><em>haplo.glm</em> fits a logistic regression model for the asthma status (<em>trait</em> - NOTE: this must be a 0/1 variable) on the inferred haplotypes (<em>genoH</em>), the names of SNPs and their allele names are passed in the <em>locus.label</em> and <em>allele.lev</em> arguments, while only haplotypes with at least 5% frequency are considered. As a result, we obtain the OR for each haplotype with their significance P-value, with respect to the most common haplotype (ATG with 44% frequency). In particular, from this analysis, we cannot see any haplotype significantly associated with asthma.</p>
</div>
<div id="sliding-window-approach" class="section level3" number="3.5.3">
<h3><span class="header-section-number">3.5.3</span> Sliding window approach</h3>
<p>The inference of haplotypes depends on a predefined region or sets of SNPs. For instance, in the last section, we selected three SNPs that were in high LD. However, when no previous knowledge is available about the region or SNPs for which haplotypes should be inferred, we can apply a sliding window for haplotype inference .</p>
<p>To illustrate this type of analysis, we now consider a second block of 10 SNPs in our asthma example (from 6th to 15th SNP). Considering large haplotypes, however, increases the number of possible haplotypes in the sample, decreasing the power of finding real associations. In addition, in predefined blocks, it is possible to miss the most efficient length of the susceptible haplotype, incrementing the loss of power. This is overcome using a sliding window. We thus ask which is the haplotype combination from any of 4, 5, 6 or 7 consecutive SNPs that gives the highest association with asthma status. We then reformat SNP genotypes in the region with the <em>make.geno</em> function and perform an association analysis for multiple haplotypes of <em>i</em> SNPs sliding from the 6th to the 15th SNP in the data. We perform an analysis for each window length <em>i</em> varying from 4 to 7 SNPs.</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="single-association-analysis.html#cb213-1" aria-hidden="true" tabindex="-1"></a>snpsH2 <span class="ot">&lt;-</span> <span class="fu">labels</span>(asthma.s)[<span class="dv">6</span><span class="sc">:</span><span class="dv">15</span>]</span>
<span id="cb213-2"><a href="single-association-analysis.html#cb213-2" aria-hidden="true" tabindex="-1"></a>genoH2 <span class="ot">&lt;-</span> <span class="fu">make.geno</span>(asthma.s, snpsH2)</span>
<span id="cb213-3"><a href="single-association-analysis.html#cb213-3" aria-hidden="true" tabindex="-1"></a>haplo.score <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb213-4"><a href="single-association-analysis.html#cb213-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">4</span><span class="sc">:</span><span class="dv">7</span>) {</span>
<span id="cb213-5"><a href="single-association-analysis.html#cb213-5" aria-hidden="true" tabindex="-1"></a> trait <span class="ot">&lt;-</span> asthma.s<span class="sc">$</span>casecontrol</span>
<span id="cb213-6"><a href="single-association-analysis.html#cb213-6" aria-hidden="true" tabindex="-1"></a> haplo.score[[i<span class="dv">-3</span>]] <span class="ot">&lt;-</span> <span class="fu">haplo.score.slide</span>(trait, genoH2, </span>
<span id="cb213-7"><a href="single-association-analysis.html#cb213-7" aria-hidden="true" tabindex="-1"></a>                          <span class="at">trait.type=</span><span class="st">&quot;binomial&quot;</span>,</span>
<span id="cb213-8"><a href="single-association-analysis.html#cb213-8" aria-hidden="true" tabindex="-1"></a>                          <span class="at">n.slide=</span>i,</span>
<span id="cb213-9"><a href="single-association-analysis.html#cb213-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">simulate=</span><span class="cn">TRUE</span>,</span>
<span id="cb213-10"><a href="single-association-analysis.html#cb213-10" aria-hidden="true" tabindex="-1"></a>                          <span class="at">sim.control=</span><span class="fu">score.sim.control</span>(<span class="at">min.sim=</span><span class="dv">100</span>,</span>
<span id="cb213-11"><a href="single-association-analysis.html#cb213-11" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">max.sim=</span><span class="dv">200</span>)) </span>
<span id="cb213-12"><a href="single-association-analysis.html#cb213-12" aria-hidden="true" tabindex="-1"></a> }</span></code></pre></div>
<p>The results can be visualized with the following plot</p>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="single-association-analysis.html#cb214-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb214-2"><a href="single-association-analysis.html#cb214-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">4</span><span class="sc">:</span><span class="dv">7</span>) {</span>
<span id="cb214-3"><a href="single-association-analysis.html#cb214-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">plot</span>(haplo.score[[i<span class="dv">-3</span>]])</span>
<span id="cb214-4"><a href="single-association-analysis.html#cb214-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">title</span>(<span class="fu">paste</span>(<span class="st">&quot;Sliding Window=&quot;</span>, i, <span class="at">sep=</span><span class="st">&quot;&quot;</span>))</span>
<span id="cb214-5"><a href="single-association-analysis.html#cb214-5" aria-hidden="true" tabindex="-1"></a> }</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:plotSliding"></span>
<img src="fig/plotSliding-1.png" alt="Sliding window approach. Results obtained of varying haplotype size from 4 up to 7 of  6th to the 15th SNP from asthma data example" width="672" />
<p class="caption">
Figure 3.6: Sliding window approach. Results obtained of varying haplotype size from 4 up to 7 of 6th to the 15th SNP from asthma data example
</p>
</div>
<p>We observe that the highest -log10(P-value) is obtained for a haplotype of 4 SNP length starting at the 4th SNP of the selected SNPs. After deciding the best combination of SNPs, the haplotype association with asthma can be estimated by</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="single-association-analysis.html#cb215-1" aria-hidden="true" tabindex="-1"></a>snpsH3 <span class="ot">&lt;-</span> snpsH2[<span class="dv">4</span><span class="sc">:</span><span class="dv">7</span>]</span>
<span id="cb215-2"><a href="single-association-analysis.html#cb215-2" aria-hidden="true" tabindex="-1"></a>genoH3 <span class="ot">&lt;-</span> <span class="fu">make.geno</span>(asthma.s, snpsH3)</span>
<span id="cb215-3"><a href="single-association-analysis.html#cb215-3" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">haplo.glm</span>(trait<span class="sc">~</span>genoH3,           </span>
<span id="cb215-4"><a href="single-association-analysis.html#cb215-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">family=</span><span class="st">&quot;binomial&quot;</span>, </span>
<span id="cb215-5"><a href="single-association-analysis.html#cb215-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">locus.label=</span>snpsH3,</span>
<span id="cb215-6"><a href="single-association-analysis.html#cb215-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">allele.lev=</span><span class="fu">attributes</span>(genoH3)<span class="sc">$</span>unique.alleles,</span>
<span id="cb215-7"><a href="single-association-analysis.html#cb215-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">control =</span> <span class="fu">haplo.glm.control</span>(<span class="at">haplo.freq.min=</span><span class="fl">0.05</span>))      </span>
<span id="cb215-8"><a href="single-association-analysis.html#cb215-8" aria-hidden="true" tabindex="-1"></a><span class="fu">intervals</span>(mod)</span></code></pre></div>
<pre><code>               freq   or   95%   C.I.   P-val 
        TCCC 0.3521   1.00 Reference haplotype 
        GCCC 0.2963   1.01 (   0.82 -   1.24 )  0.9495 
        TTCA 0.1184   0.89 (   0.67 -   1.20 )  0.4433 
        TTTA 0.1830   1.12 (   0.87 -   1.43 )  0.3796 
 genoH3.rare 0.0502   0.78 (   0.50 -   1.22 )  0.2750 </code></pre>
<p>Here, we observe that individuals carrying the haplotype GCAC have a 53% increased risk of asthma relative to those having the reference haplotype TCGT (OR=1.53, p=0.0021). The haplotype GTCA is also significantly associated with the disease (p=0.0379). A likelihood ratio test for haplotype status can be extracted from the results of the <em>haplo.glme</em> function:</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="single-association-analysis.html#cb217-1" aria-hidden="true" tabindex="-1"></a>lrt <span class="ot">&lt;-</span> mod<span class="sc">$</span>lrt</span>
<span id="cb217-2"><a href="single-association-analysis.html#cb217-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pchisq</span>(lrt<span class="sc">$</span>lrt, lrt<span class="sc">$</span>df, <span class="at">lower=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>[1] 0.5044641</code></pre>
<p>We can also test the association between asthma and the haplotype adjusted for smoking status</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="single-association-analysis.html#cb219-1" aria-hidden="true" tabindex="-1"></a>smoke <span class="ot">&lt;-</span> asthma.s<span class="sc">$</span>smoke</span>
<span id="cb219-2"><a href="single-association-analysis.html#cb219-2" aria-hidden="true" tabindex="-1"></a>mod.adj.ref <span class="ot">&lt;-</span> <span class="fu">glm</span>(trait <span class="sc">~</span> smoke, <span class="at">family=</span><span class="st">&quot;binomial&quot;</span>)</span>
<span id="cb219-3"><a href="single-association-analysis.html#cb219-3" aria-hidden="true" tabindex="-1"></a>mod.adj <span class="ot">&lt;-</span> <span class="fu">haplo.glm</span>(trait <span class="sc">~</span> genoH3 <span class="sc">+</span> smoke ,           </span>
<span id="cb219-4"><a href="single-association-analysis.html#cb219-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">family=</span><span class="st">&quot;binomial&quot;</span>, </span>
<span id="cb219-5"><a href="single-association-analysis.html#cb219-5" aria-hidden="true" tabindex="-1"></a>                 <span class="at">locus.label=</span>snpsH3,</span>
<span id="cb219-6"><a href="single-association-analysis.html#cb219-6" aria-hidden="true" tabindex="-1"></a>                 <span class="at">allele.lev=</span><span class="fu">attributes</span>(genoH3)<span class="sc">$</span>unique.alleles,</span>
<span id="cb219-7"><a href="single-association-analysis.html#cb219-7" aria-hidden="true" tabindex="-1"></a>                 <span class="at">control =</span> <span class="fu">haplo.glm.control</span>(<span class="at">haplo.freq.min=</span><span class="fl">0.05</span>))</span>
<span id="cb219-8"><a href="single-association-analysis.html#cb219-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb219-9"><a href="single-association-analysis.html#cb219-9" aria-hidden="true" tabindex="-1"></a>lrt.adj <span class="ot">&lt;-</span> mod.adj.ref<span class="sc">$</span>deviance <span class="sc">-</span> mod.adj<span class="sc">$</span>deviance</span>
<span id="cb219-10"><a href="single-association-analysis.html#cb219-10" aria-hidden="true" tabindex="-1"></a><span class="fu">pchisq</span>(lrt.adj, mod.adj<span class="sc">$</span>lrt<span class="sc">$</span>df, <span class="at">lower=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<pre><code>[1] 0.6497138</code></pre>
</div>
</div>
<div id="genetic-score" class="section level2" number="3.6">
<h2><span class="header-section-number">3.6</span> Genetic score</h2>
<p>A genetic score, also called a polygenic risk score or genetic risk score, is given by the number of risk alleles that an individual carries <span class="citation">(<a href="#ref-dudbridge2013power" role="doc-biblioref">Dudbridge 2013</a>)</span>. Genetic scores aim to assess the collective prediction of the phenotype risk associated with multiple SNPs. The SNPs to be included in a genetic score can be selected from different sources. They may be selected from candidate genes, from reported SNPs associated with the trait, or from significant SNPs from a given analysis.</p>
<p>We illustrate how to perform a genetic risk score analysis with the asthma study, using the package <code>PredictABEL</code>. We start by performing the single SNP analysis, as previously discussed</p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="single-association-analysis.html#cb221-1" aria-hidden="true" tabindex="-1"></a>ans <span class="ot">&lt;-</span> <span class="fu">WGassociation</span>(casecontrol, asthma.s, <span class="at">model=</span><span class="st">&quot;log-add&quot;</span>)</span></code></pre></div>
<p>We choose the SNPs that show an association at 0.1 level to be included in a multivariate model,</p>
<div class="sourceCode" id="cb222"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb222-1"><a href="single-association-analysis.html#cb222-1" aria-hidden="true" tabindex="-1"></a>sel <span class="ot">&lt;-</span> <span class="fu">labels</span>(asthma.s)[<span class="fu">additive</span>(ans)<span class="sc">&lt;</span><span class="fl">0.1</span>]</span>
<span id="cb222-2"><a href="single-association-analysis.html#cb222-2" aria-hidden="true" tabindex="-1"></a>asthma.sel <span class="ot">&lt;-</span> asthma.s[,sel]</span>
<span id="cb222-3"><a href="single-association-analysis.html#cb222-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(asthma.sel)</span></code></pre></div>
<pre><code>  rs1422993 rs184448 rs324957 rs324960 rs324981 rs727162 rs6084432
1       G/G      T/T      G/G      C/T      A/T      G/G       G/G
2       G/T      G/G      A/A      C/C      A/A      G/G       G/G
3       G/G      T/T      G/G      C/T      T/T      G/C       G/A
4       G/T      T/G      G/A      C/C      A/T      G/G       G/G
5       G/T      T/T      G/G      C/C      A/T      G/G       G/G
6       G/G      T/T      G/G      C/T      A/T      G/C       G/G</code></pre>
<p>We join the SNPs and case/control variable in a single data.frame. Using the <code>additive()</code>, we recode the SNP genotypes as 0, 1 or 2 where the homozygous for the alternative allele is 2. As such, the genetic score is the sum of alternative alleles over selected SNPs</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="single-association-analysis.html#cb224-1" aria-hidden="true" tabindex="-1"></a>asthma.sel <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">lapply</span>(asthma.sel, additive))</span>
<span id="cb224-2"><a href="single-association-analysis.html#cb224-2" aria-hidden="true" tabindex="-1"></a>dd.end <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">casecontrol=</span>asthma.s<span class="sc">$</span>casecontrol, asthma.sel)</span>
<span id="cb224-3"><a href="single-association-analysis.html#cb224-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(dd.end)</span></code></pre></div>
<pre><code>  casecontrol rs1422993 rs184448 rs324957 rs324960 rs324981 rs727162
1           0         0        0        0        1        1        0
2           0         1        2        2        0        0        0
3           0         0        0        0        1        2        1
4           0         1        1        1        0        1        0
5           1         1        0        0        0        1        0
6           0         0        0        0        1        1        1
  rs6084432
1         0
2         0
3         1
4         0
5         0
6         0</code></pre>
<p>A multivariate regression model is fitted with all the SNPs to further select those that best predict the asthma status. An automatic variable selection, based on Akaike information criteria (AIC), can be applied to the model using the <code>stepAIC()</code> function from library <code>MASS</code>. Note that the function does not support missing values and requires having complete cases (i.e. no missing values on genotypes).</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="single-association-analysis.html#cb226-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MASS)</span>
<span id="cb226-2"><a href="single-association-analysis.html#cb226-2" aria-hidden="true" tabindex="-1"></a>dd.end.complete <span class="ot">&lt;-</span> dd.end[<span class="fu">complete.cases</span>(dd.end),]</span>
<span id="cb226-3"><a href="single-association-analysis.html#cb226-3" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> <span class="fu">stepAIC</span>(<span class="fu">glm</span>(casecontrol <span class="sc">~</span> ., dd.end.complete,</span>
<span id="cb226-4"><a href="single-association-analysis.html#cb226-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">family=</span><span class="st">&quot;binomial&quot;</span>),</span>
<span id="cb226-5"><a href="single-association-analysis.html#cb226-5" aria-hidden="true" tabindex="-1"></a>               <span class="at">method=</span><span class="st">&quot;forward&quot;</span>, <span class="at">trace=</span><span class="dv">0</span>)</span></code></pre></div>
<p>The selected SNPs, from the multivariate model, in <code>snps.score</code> are then used to create the genetic score. First, a summary shows the SNPs’ associations of the selected SNPs with the trait</p>
<div class="sourceCode" id="cb227"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb227-1"><a href="single-association-analysis.html#cb227-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mod)</span></code></pre></div>
<pre><code>
Call:
glm(formula = casecontrol ~ rs1422993 + rs324957 + rs727162 + 
    rs6084432, family = &quot;binomial&quot;, data = dd.end.complete)

Deviance Residuals: 
    Min       1Q   Median       3Q      Max  
-0.9775  -0.7290  -0.6655  -0.5596   1.9655  

Coefficients:
            Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept) -1.77498    0.13635 -13.018  &lt; 2e-16 ***
rs1422993    0.20528    0.09849   2.084  0.03714 *  
rs324957     0.25937    0.09266   2.799  0.00512 ** 
rs727162     0.18037    0.10447   1.727  0.08424 .  
rs6084432    0.19982    0.11388   1.755  0.07932 .  
---
Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 1587.0  on 1517  degrees of freedom
Residual deviance: 1569.2  on 1513  degrees of freedom
AIC: 1579.2

Number of Fisher Scoring iterations: 4</code></pre>
<p>This is the code to retrieve the selected SNPs to create the genetic score:</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="single-association-analysis.html#cb229-1" aria-hidden="true" tabindex="-1"></a>snps.score <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">coef</span>(mod))[<span class="sc">-</span><span class="dv">1</span>]</span>
<span id="cb229-2"><a href="single-association-analysis.html#cb229-2" aria-hidden="true" tabindex="-1"></a>snps.score</span></code></pre></div>
<pre><code>[1] &quot;rs1422993&quot; &quot;rs324957&quot;  &quot;rs727162&quot;  &quot;rs6084432&quot;</code></pre>
<p>The function <code>riskScore()</code> of the <code>PredictABEL</code> package creates the risk score</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="single-association-analysis.html#cb231-1" aria-hidden="true" tabindex="-1"></a>pos <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">names</span>(dd.end.complete)<span class="sc">%in%</span>snps.score)</span>
<span id="cb231-2"><a href="single-association-analysis.html#cb231-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(dd.end.complete)</span></code></pre></div>
<pre><code>[1] &quot;casecontrol&quot; &quot;rs1422993&quot;   &quot;rs184448&quot;    &quot;rs324957&quot;   
[5] &quot;rs324960&quot;    &quot;rs324981&quot;    &quot;rs727162&quot;    &quot;rs6084432&quot;  </code></pre>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="single-association-analysis.html#cb233-1" aria-hidden="true" tabindex="-1"></a>pos</span></code></pre></div>
<pre><code>[1] 2 4 7 8</code></pre>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="single-association-analysis.html#cb235-1" aria-hidden="true" tabindex="-1"></a>score <span class="ot">&lt;-</span> <span class="fu">riskScore</span>(mod, <span class="at">data=</span>dd.end.complete, </span>
<span id="cb235-2"><a href="single-association-analysis.html#cb235-2" aria-hidden="true" tabindex="-1"></a>                      <span class="at">cGenPreds=</span>pos,</span>
<span id="cb235-3"><a href="single-association-analysis.html#cb235-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">Type=</span><span class="st">&quot;unweighted&quot;</span>)</span>
<span id="cb235-4"><a href="single-association-analysis.html#cb235-4" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(score)</span></code></pre></div>
<pre><code>score
  0   1   2   3   4   5   6 
101 389 491 347 139  44   7 </code></pre>
<p>The distribution of the score across individuals can be seen in Figure <a href="single-association-analysis.html#fig:histGS">3.7</a></p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="single-association-analysis.html#cb237-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(score, <span class="at">col=</span><span class="st">&quot;gray90&quot;</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:histGS"></span>
<img src="fig/histGS-1.png" alt="Distribution of the genetic score used to predict case/control status in the asthma example." width="672" />
<p class="caption">
Figure 3.7: Distribution of the genetic score used to predict case/control status in the asthma example.
</p>
</div>
<p>Once the genetic score is created, we test its association with asthma status using a general linear model</p>
<div class="sourceCode" id="cb238"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb238-1"><a href="single-association-analysis.html#cb238-1" aria-hidden="true" tabindex="-1"></a>mod.lin <span class="ot">&lt;-</span> <span class="fu">glm</span>(casecontrol<span class="sc">~</span>score, dd.end.complete,</span>
<span id="cb238-2"><a href="single-association-analysis.html#cb238-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">family=</span><span class="st">&quot;binomial&quot;</span>)</span>
<span id="cb238-3"><a href="single-association-analysis.html#cb238-3" aria-hidden="true" tabindex="-1"></a>mod.lin</span></code></pre></div>
<pre><code>
Call:  glm(formula = casecontrol ~ score, family = &quot;binomial&quot;, data = dd.end.complete)

Coefficients:
(Intercept)        score  
    -1.7612       0.2151  

Degrees of Freedom: 1517 Total (i.e. Null);  1516 Residual
Null Deviance:      1587 
Residual Deviance: 1570     AIC: 1574</code></pre>
<div class="sourceCode" id="cb240"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb240-1"><a href="single-association-analysis.html#cb240-1" aria-hidden="true" tabindex="-1"></a><span class="fu">exp</span>(<span class="fu">coef</span>(mod.lin)[<span class="dv">2</span>])</span></code></pre></div>
<pre><code>   score 
1.240039 </code></pre>
<p>We observe that the risk of asthma increases 24% per each risk allele. The predictive power of the genetic score can be assessed by computing the area under the ROC curve (AUC).</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="single-association-analysis.html#cb242-1" aria-hidden="true" tabindex="-1"></a>predrisk <span class="ot">&lt;-</span> <span class="fu">predRisk</span>(mod.lin, dd.end.complete)</span>
<span id="cb242-2"><a href="single-association-analysis.html#cb242-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotROC</span>(<span class="at">data=</span>dd.end.complete, <span class="at">cOutcome=</span><span class="dv">1</span>,</span>
<span id="cb242-3"><a href="single-association-analysis.html#cb242-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">predrisk =</span> predrisk)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:rocGS"></span>
<img src="fig/rocGS-1.png" alt="ROC curve of the genetic score used to predict case/control status in the asthma example." width="672" />
<p class="caption">
Figure 3.8: ROC curve of the genetic score used to predict case/control status in the asthma example.
</p>
</div>
<pre><code>AUC [95% CI] for the model 1 :  0.574 [ 0.54  -  0.607 ] </code></pre>
<p>Figure <a href="single-association-analysis.html#fig:rocGS">3.8</a> shows that the predictive power of the genetic score is 57.4% with a 95% confidence interval (54.0 - 60.7).</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-dudbridge2013power" class="csl-entry">
Dudbridge, Frank. 2013. <span>“Power and Predictive Accuracy of Polygenic Risk Scores.”</span> <em>PLoS Genetics</em> 9 (3): e1003348.
</div>
<div id="ref-gonzalez2008maximizing" class="csl-entry">
González, Juan R, Josep L Carrasco, Frank Dudbridge, Lluı́s Armengol, Xavier Estivill, and Victor Moreno. 2008. <span>“Maximizing Association Statistics over Genetic Models.”</span> <em>Genetic Epidemiology</em> 32 (3): 246–54.
</div>
<div id="ref-wigginton2005note" class="csl-entry">
Wigginton, Janis E, David J Cutler, and Gonçalo R Abecasis. 2005. <span>“A Note on Exact Tests of <span>H</span>ardy-<span>W</span>einberg Equilibrium.”</span> <em>The American Journal of Human Genetics</em> 76 (5): 887–93.
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="bioconductor.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="genome-wide-association-studies.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/isglobal-brge/Master_Bioinformatics/tree/master/docs01-Genomic_data_analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "lunr",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
